-- MySQL dump 10.13  Distrib 5.1.62, for unknown-linux-gnu (x86_64)
--
-- Host: ens-staging2    Database: rattus_norvegicus_core_85_6
-- ------------------------------------------------------
-- Server version	5.1.58

/*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
/*!40101 SET @OLD_CHARACTER_SET_RESULTS=@@CHARACTER_SET_RESULTS */;
/*!40101 SET @OLD_COLLATION_CONNECTION=@@COLLATION_CONNECTION */;
/*!40101 SET NAMES utf8 */;
/*!40103 SET @OLD_TIME_ZONE=@@TIME_ZONE */;
/*!40103 SET TIME_ZONE='+00:00' */;
/*!40014 SET @OLD_UNIQUE_CHECKS=@@UNIQUE_CHECKS, UNIQUE_CHECKS=0 */;
/*!40014 SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0 */;
/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
/*!40111 SET @OLD_SQL_NOTES=@@SQL_NOTES, SQL_NOTES=0 */;

--
-- Table structure for table `attrib_type`
--

DROP TABLE IF EXISTS `attrib_type`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `attrib_type` (
  `attrib_type_id` smallint(5) unsigned NOT NULL AUTO_INCREMENT,
  `code` varchar(20) NOT NULL DEFAULT '',
  `name` varchar(255) NOT NULL DEFAULT '',
  `description` text,
  PRIMARY KEY (`attrib_type_id`),
  UNIQUE KEY `code_idx` (`code`)
) ENGINE=MyISAM AUTO_INCREMENT=518 DEFAULT CHARSET=latin1;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `attrib_type`
--

LOCK TABLES `attrib_type` WRITE;
/*!40000 ALTER TABLE `attrib_type` DISABLE KEYS */;
INSERT INTO `attrib_type` VALUES (1,'embl_acc','European Nucleotide Archive (was EMBL) accession',NULL),(2,'status','Status',NULL),(3,'synonym','Synonym',NULL),(4,'name','Name','Alternative/long name'),(5,'type','Type of feature',NULL),(6,'toplevel','Top Level','Top Level Non-Redundant Sequence Region'),(7,'GeneCount','Gene Count','Total Number of Genes'),(10,'SNPCount','Short Variants','Total Number of SNPs'),(11,'codon_table','Codon Table','Alternate codon table'),(12,'_selenocysteine','Selenocysteine',NULL),(13,'bacend','bacend',NULL),(14,'htg','htg','High Throughput phase attribute'),(15,'miRNA','Micro RNA','Coordinates of the mature miRNA'),(16,'non_ref','Non Reference','Non Reference Sequence Region'),(17,'sanger_project','Sanger Project name',NULL),(18,'clone_name','Clone name',NULL),(19,'fish','FISH location',NULL),(21,'org','Sequencing centre',NULL),(22,'method','Method',NULL),(23,'superctg','Super contig id',NULL),(24,'inner_start','Max start value',NULL),(25,'inner_end','Min end value',NULL),(26,'state','Current state of clone',NULL),(27,'organisation','Organisation sequencing clone',NULL),(28,'seq_len','Accession length',NULL),(29,'fp_size','FP size',NULL),(30,'BACend_flag','BAC end flags',NULL),(31,'fpc_clone_id','fpc clone',NULL),(32,'KnwnPCCount','protein_coding_KNOWN','Number of Known Protein Coding'),(33,'NovPCCount','protein_coding_NOVEL','Number of Novel Protein Coding'),(36,'PredPCCount','protein_coding_PREDICTED','Number of Predicted Protein Coding'),(37,'IGGeneCount','IG_gene','Number of IG Genes'),(38,'IGPsGenCount','IG_pseudogene','Number of IG Pseudogenes'),(39,'TotPsCount','total_pseudogene','Total Number of Pseudogenes'),(42,'KnwnPCProgCount','protein_coding_in_progress_KNOWN','Number of Known Protein Coding in progress'),(43,'NovPCProgCount','protein_coding_in_progress_NOVEL','Number of Novel Protein Coding in progress'),(44,'AnnotSeqLength','Annotated sequence length','Annotated Sequence'),(45,'TotCloneNum','Total number of clones','Total Number of Clones'),(46,'NumAnnotClone','Fully annotated clones','Number of Fully Annotated Clones'),(47,'ack','Acknowledgement','Acknowledgement for manual annotation'),(48,'htg_phase','High throughput phase','High throughput genomic sequencing phase'),(49,'description','Description','A general descriptive text attribute'),(50,'chromosome','Chromosome','Chromosomal location for supercontigs that are not assembled'),(51,'nonsense','Nonsense Mutation','Strain specific nonesense mutation'),(52,'author','Author','Group resonsible for Vega annotation'),(53,'author_email','Author email address','Author email address'),(54,'remark','Remark','Annotation remark'),(55,'transcr_class','Transcript class','Transcript class'),(57,'ccds','CCDS','CCDS identifier'),(58,'CCDS_PublicNote','CCDS Public Note','Public Note for CCDS identifier, provided by http://www.ncbi.nlm.nih.gov/CCDS'),(59,'Frameshift','Frameshift','Frameshift modelled as intron'),(62,'ncRNA','Structure','RNA secondary structure line'),(63,'skip_clone','skip clone  Skip clone in align_by_clone_identity.pl',NULL),(64,'coding_cnt','Coding genes','Number of protein coding Genes'),(67,'pseudogene_cnt','Pseudogenes','Number of pseudogenes'),(80,'supercontig','SuperContig name',NULL),(81,'well_name','Well plate name',NULL),(82,'bacterial','Bacterial',NULL),(90,'bacend_well_nam','BACend well name',NULL),(91,'alt_well_name','Alt well name',NULL),(92,'TranscriptEdge','Transcript Edge',NULL),(93,'alt_embl_acc','Alt European Nucleotide Archive (was EMBL) acc',NULL),(94,'alt_org','Alt org',NULL),(95,'intl_clone_name','International Clone Name',NULL),(96,'embl_version','European Nucleotide Archive (was EMBL) Version',NULL),(97,'chr','Chromosome Name','Chromosome Name Contained in the Assembly'),(98,'equiv_asm','Equivalent EnsEMBL assembly','For full chromosomes made from NCBI AGPs'),(109,'HitSimilarity','hit similarity','percentage id to parent transcripts'),(110,'HitCoverage','hit coverage','coverage of parent transcripts'),(111,'PropNonGap','proportion non gap','proportion non gap'),(112,'NumStops','number of stops',NULL),(113,'GapExons','gap exons','number of gap exons'),(114,'SourceTran','source transcript','source transcript'),(115,'EndNotFound','end not found','end not found'),(116,'StartNotFound','start not found','start not found'),(117,'Frameshift Fra','Frameshift modelled as intron',NULL),(118,'ensembl_name','Ensembl name','Name of equivalent Ensembl chromosome'),(119,'NoAnnotation','NoAnnotation','Clones without manual annotation'),(120,'hap_contig','Haplotype contig','Contig present on a haplotype'),(121,'annotated','Clone Annotation Status',NULL),(122,'keyword','Clone Keyword',NULL),(123,'hidden_remark','Hidden Remark',NULL),(124,'mRNA_start_NF','mRNA start not found',NULL),(125,'mRNA_end_NF','mRNA end not found',NULL),(126,'cds_start_NF','CDS start not found',NULL),(127,'cds_end_NF','CDS end not found',NULL),(128,'write_access','Write access for Sequence Set','1 for writable , 0 for read-only'),(129,'hidden','Hidden Sequence Set',NULL),(130,'vega_name','Vega name','Vega seq_region.name'),(131,'vega_export_mod','Export mode','E (External), I (Internal) etc'),(132,'vega_release','Vega release','Vega release number'),(133,'atag_CLE','Clone_left_end','Clone_lef_end feature marked in GAP database'),(134,'atag_CRE','Clone_right_end','Clone_right_end feature marked in GAP database'),(135,'atag_Misc','Misc','miscellaneous feature marked in GAP database'),(136,'atag_Unsure','Unsure','region of uncertain DNA sequence marked in GAP database'),(137,'MultAssem','Multiple Assembled seq region','Part of Seq Region is part of more than one assembly'),(140,'wgs','WGS contig','WGS contig integrated into the map'),(141,'bac','AGP clones','tiling path of clones'),(142,'GeneGC','Gene GC','Percentage GC content for this gene'),(143,'TotAssemblyLeng','Finished sequence length','Length of the assembly not counting sequence gaps'),(144,'amino_acid_sub','Amino acid substitution','Some translations have been manually curated for amino acid substitiutions. For example a stop codon may be changed to an amino acid in order to prevent premature truncation, or one amino acid can be substituted for another.'),(145,'_rna_edit','rna_edit','RNA edit'),(146,'kill_reason','Kill Reason','Reason why a transcript has been killed'),(147,'strip_UTR','Strip UTR','Transcript needs bad UTR removing'),(148,'TotAssLength','Finished sequence length','Finished Sequence'),(149,'PsCount','pseudogene','Number of Pseudogenes'),(152,'TotPTCount','total_processed_transcript','Total Number of Processed Transcripts'),(153,'TotPCCount','total_protein_coding','Total Number of Protein Coding'),(156,'PolyPsCount','polymorphic_pseudogene','Number of Polymorphic Pseudogenes'),(157,'TotIGGeneCount','total_IG_gene','Total Number of IG Genes'),(158,'ProcPsCount','proc_pseudogene','Number of Processed Pseudogenes'),(159,'UnPsCount','unproc_pseudogene','Number of Unprocessed Pseudogenes'),(160,'TPsCount','transcribed_pseudogene','Number of Transcribed Pseudogenes'),(161,'TECCount','TEC','Number of TEC Genes'),(164,'IsoPoint','Isoelectric point','Pepstats attributes'),(165,'Charge','Charge','Pepstats attributes'),(166,'MolecularWeight','Molecular weight','Pepstats attributes'),(167,'NumResidues','Number of residues','Pepstats attributes'),(168,'AvgResWeight','Ave. residue weight','Pepstats attributes'),(170,'initial_met','Initial methionine','Set first amino acid to methionine'),(171,'NonGapHCov','NonGapHCov',NULL),(172,'otter_support','otter support','Evidence ID that was used as supporting feature for building a gene in Vega'),(173,'enst_link','enst link','Code to link a OTTT with an ENST when they both share the CDS of ENST'),(174,'upstream_ATG','upstream ATG','Alternative ATG found upstream of the defined as start ATG for the transcript'),(175,'TPPsCount','transcribed_processed_pseudogene','Number of Transcribed Processed Pseudogenes'),(176,'TUPsCount','transcribed_unprocessed_pseudogene','Number of Transcribed Unprocessed Pseudogenes'),(177,'UniPsCount','unitary_pseudogene','Number of Unitary Pseudogenes'),(180,'TUyPsCount','transcribed_unitary_pseudogene','Number of Transcribed Unitary Pseudogenes'),(181,'PolyCount','polymorphic','Number of Polymorphic Genes'),(184,'TRGeneCount','TR_gene','Number of TR Genes'),(185,'TRPsCount','TR_pseudo','Number of TR Pseudogenes'),(186,'tp_ott_support','otter protein transcript support','Evidence ID that was used as supporting feature for building a gene in Vega'),(187,'td_ott_support','otter dna transcript support','Evidence ID that was used as supporting feature for building a gene in Vega'),(188,'ep_ott_support','otter protein exon support','Evidence ID that was used as supporting feature for building a gene in Vega'),(189,'ed_ott_support','otter dna exon support','Evidence ID that was used as supporting feature for building a gene in Vega'),(191,'StopGained','SNP causes stop codon to be gained','This transcript has a variant that causes a stop codon to be gained in at least 10 percent of a HapMap population'),(192,'StopLost','SNP causes stop codon to be lost','This transcript has a variant that causes a stop codon to be lost in at least 10 percent of a HapMap population'),(198,'lost_frameshift','lost_frameshift','Frameshift on the query sequence is lost in the target sequence'),(199,'AltThreePrime','Alternate three prime end','The position of other possible three prime ends for the transcript'),(216,'GeneInLRG','Gene in LRG','This gene is contained within an LRG region'),(217,'GeneOverlapLRG','Gene overlaps LRG','This gene is partially overlapped by a LRG region (start or end outside LRG)'),(218,'readthrough_tra','readthrough transcript','Havana readthrough transcripts'),(300,'CNE','Constitutive exon','An exon that is always included in the mature mRNA, even in different mRNA isoforms'),(301,'CE','Cassette exon','One exon is spliced out of the primary transcript together with its flanking introns'),(302,'IR','Intron retention','A sequence is spliced out as an intron or remains in the mature mRNA transcript'),(303,'MXE','Mutually exclusive exons','In the simpliest case, one or two consecutive exons are retained but not both'),(304,'A3SS','Alternative 3\' sites','Two or more splice sites are recognized at the 5\' end of an exon. An alternative 3\' splice junction (acceptor site) is used, changing the 5\' boundary of the downstream exon'),(305,'A5SS','Alternative 5\' sites','Two or more splice sites are recognized at the 3\' end of an exon. An alternative 5\' splice junction (donor site) is used, changing the 3\' boundary of the upstream exon'),(306,'AFE','Alternative first exon','The second exons of each variant have identical boundaries, but the first exons do not overlap'),(307,'ALE','Alternative last exon','Penultimate exons of each splice variant have identical boundaries, but the last exons do not overlap'),(308,'II','Intron isoform','Alternative donor or acceptor splice sites lead to truncation or extension of introns, respectively'),(309,'EI','Exon isoform','Alternative donor or acceptor splice sites leads to truncation or extension of exons, respectively'),(310,'AI','Alternative initiation','Alternative choice of promoters'),(311,'AT','Alternative termination','Alternative choice of polyadenylation sites'),(312,'patch_fix','Assembly Patch Fix','Assembly patch that will, in the next assembly release, replace the corresponding sequence found in the current assembly'),(313,'patch_novel','Assembly Patch Novel','Assembly patch that will, in the next assembly release, be retained as an alternate non-reference sequence in a similar way to haplotypes'),(314,'LRG','Locus Reference Genomic','Locus Reference Genomic sequence'),(315,'NoEvidence','Evidence for transcript removed','Supporting evidence for this projected transcript has been removed'),(316,'circular_seq','Circular sequence','Circular chromosome or plasmid molecule'),(317,'external_db','External database','External database to which seq_region name may be linked'),(318,'split_tscript','split_tscript','split_tscript'),(319,'Threep','Three prime end','Alternate three prime end'),(320,'gene_cluster','Gene cluster','Havana annotated gene cluster'),(328,'_rib_frameshift','Ribosomal Frameshift','Position and magnitude of frameshift'),(345,'vega_ref_chrom','Vega reference chromosome','Haplotypes reference a regular chromosome (indicated in the value of the attribute)'),(346,'PutPCCount','protein_coding_PUTATIVE','Number of Putative Protein Coding'),(347,'proj_alt_seq','Projection altered sequence','Projected sequence differs from original'),(348,'hav_gene_type','Havana gene biotype','Gene biotype assigned by Havana'),(353,'noncoding_cnt','Non coding genes','Number of non coding genes'),(358,'PHIbase_mutant','PHI-base mutant','PHI-base phenotype of the mutants'),(360,'ncrna_host','ncrna_host','Havana ncrna_host gene'),(361,'peptide-class','Peptide classification','The classification of the gene or transcript based on alignment to NR (values: TE WH NH)'),(362,'working-set','Working Gene Set','High-confidence set of genes, composed of evidence-based genes and non-overlapping protein-coding ab initio gene models'),(363,'filtered-set','Filtered Gene Set v1','Working genes that are screened for TE content and orthology with sorghum and rice'),(364,'super-set','Super Working Gene Set','Set of all working gene set loci from both Builds 4a and 5a'),(365,'projected4a2','Projected by alignment','Temporary (Monday, August 23, 2010)'),(366,'merged','Merged species',NULL),(367,'karyotype_rank','Rank in the karyotype','For a given seq_region, if it is part of the species karyotype, will indicate its rank'),(368,'noncoding_acnt','Non coding genes','Number of non coding genes on alternate sequences'),(369,'coding_acnt','Coding genes','Number of protein coding genes on alternate sequences'),(370,'pseudogene_acnt','Pseudogenes','Number of pseudogenes on alternate sequences'),(371,'clone_end','Clone end','Side of the contig on which a vector lies (enum:RIGHT, LEFT)'),(372,'contig_scaffold','Contig Scaffold','Scaffold that contains mutually ordered contigs'),(373,'current_version','Current Accession Version','Identifies the most recent version of an accession'),(374,'seq_status','Sequence Status','Sequence status.'),(375,'clone_vector','Vector sequence','A clone-end vector associated with a contig (enum:SP6, T7).'),(376,'creation_date','Creation date','Creation date of annotation'),(377,'update_date','Update date','Last update date of annotation'),(378,'seq_date','Sequence date','Sequence date'),(379,'has_stop_codon','Contains stop codon','Translation attribute'),(380,'havana_cv','Havana CV term','Controlled vocabulary terms from Havana'),(381,'TlPPsCount','translated_processed_pseudogene','Number of Translated Processed Pseudogenes'),(382,'NoTransRefError','No translations due to reference error','This gene is believed to include protein coding transcripts, but no transcript has a translation due to a reference assembly error making specifying the translation impossible.'),(383,'parent_exon_key','parent_exon_key','The exon key to identify a projected transcript\'s parent transcript.'),(386,'parent_sid','parent_sid','The parent stable ID to identify a projected transcript\'s parent transcript. For internal statistics use only since this method does not work in all cases.'),(387,'noncoding_acnt_s','Small non coding genes','Number of small non coding genes on alternate sequences'),(388,'noncoding_acnt_m','Misc non coding genes','Number of unclassified (miscellaneous) non coding genes on alternate sequences'),(389,'noncoding_cnt_s','Small non coding genes','Number of small non coding genes'),(390,'noncoding_cnt_l','Long non coding genes','Number of long non coding genes'),(391,'TlUPsCount','translated_unprocessed_pseudogene','Number of Translated Unprocessed Pseudogenes'),(393,'AFFYMETRIXCount','AFFYMETRIX Count','Total Number of AFFYMETRIX features'),(394,'RFLPCount','RFLP Count','Total Number of RFLP features'),(395,'xref_id','Xref ID','ID of associated database reference'),(396,'vega_chr_type','Vega chrom type','Type of chromosome - haplotype, other, etc'),(398,'genscan','Genscan gene predictions','Number of prediction genes generated by Genscan'),(399,'gsc','GSC gene prediction','Number of prediction genes generated by gsc'),(400,'snap','Snap gene prediction','Number of prediction genes generated by Snap'),(401,'fgenesh','FGENESH gene prediction','Number of prediction genes generated by FGENESH'),(402,'genefinder','Genefinder gene prediction','Number of prediction genes generated by Genefinder'),(403,'transcript_cnt','Gene transcripts','Number of transcripts'),(404,'transcript_acnt','Gene transcripts','Number of transcripts on the alternate sequences'),(405,'ref_length','Golden Path Length','Length of the primary assembly'),(406,'total_length','Base Pairs','Total length of the assembly'),(407,'refseq_compare','refseq_compare','This attribute can be applied to both gene and transcript. It is supposed to give an indication of whether the annotation in the ensembl database is matched by annotation that we have imported from refseq. At the gene level, the match is unlikely to be an exact match because all or some of the transcripts may differ. Also, the biotype e.g. coding potential may differ. therefore, matching is a bit fuzzy and is done primarily on genomic location and then also takes gene length and gene name into consideration.'),(408,'coding_rcnt','Readthrough coding genes','Number of readthrough coding genes'),(409,'coding_racnt','Readthrough coding genes','Number of readthrough coding genes on alternate sequences'),(410,'noncoding_racnt_l','Readthrough long non coding genes','Number of readthrough long non coding genes on alternate sequences'),(411,'noncoding_racnt_s','Readthrough small non coding genes','Number of readthrough small non coding genes on alternate sequences'),(412,'noncoding_rcnt_s','Readthrough small non coding genes','Number of readthrough small non coding genes'),(413,'noncoding_rcnt_l','Readthrough long non coding genes','Number of readthrough long non coding genes'),(414,'pseudogene_rcnt','Readthrough pseudogenes','Number of readthrough pseudogenes'),(415,'pseudogene_racnt','Readthrough pseudogenes','Number of readthrough pseudogenes on alternate sequences'),(416,'gencode_level','GENCODE annotation level','level 1 (verified loci), level 2 (manually annotated loci), level 3 (automatically annotated loci)'),(417,'gencode_basic','GENCODE basic annotation','GENCODE Basic is a view provided by UCSC for users. It includes a subset of the GENCODE transcripts. In general, for protein coding genes it will show only the full length models (unless a protein coding gene has no full-length models, in which case other rules apply). For noncoding genes, it will also only show the full-length (mRNA start and end found) models (unless there are no full-length models, in which case other rules apply).'),(418,'struct_var','Structural variants','Total Number of structural variants'),(419,'genblast','GenBlastG gene predictions','Number of prediction genes generated by GenBlastG'),(420,'syn_gene_pairs','Syntenic gene pairs','Syntenic gene relationship from Gramene pipeline'),(421,'vectorbase_maker_pre','VectorBase gene predictions','Number of prediction genes generated with MAKER, by VectorBase.'),(422,'trnascan','tRNAscan-SE predictions','Number of predicted tRNA genes generated by tRNAscan-SE'),(423,'tgac_pred_supp7','T. turgidum RNA-seq alignments','Number of T. turgidum RNA-seq alignments from Krasileva et al.'),(424,'tgac_pred_supp17','T. aestivum RNA-seq alignments','Number of T. aestivum RNA-seq alignments from Krasileva et al.'),(425,'genome_component','Genome Component Name','For polyploid genome, the genome component name the seq_region belongs to'),(426,'transcript_whl','RNA-seq transcripts','RNA-seq transcripts from EchinoBase'),(427,'appris','APPRIS','APPRIS is a system that deploys a range of computational methods to provide value to the annotations of the human genome. APPRIS also selects one of the CDS for each gene as the principal isoform. APPRIS defines the principal variant by combining protein structural and functional information and information from the conservation of related species. principal1 - APPRIS principal isoform. principal2 - APPRIS candidate isoform (CCDS). principal3 - APPRIS candidate isoform (earliest CCDS). principal4 - APPRIS candidate isoform (longest CCDS). principal5 - APPRIS candidate isoform (longest coding sequence). alternative1 - APPRIS candidate isoform that is conserved in at least three tested species. alternative2 - APPRIS candidate isoform that appears to be conserved in fewer than three tested species'),(428,'TSL','Transcript Support Level','Transcription Support Level (TSL) is a method to highlight the well-supported and poorly-supported transcript models for users. The method relies on the primary data that can support full-length transcript structure and data are provided by UCSC.  The following categories are assigned to each of the evaluated annotations. tsl1 - all splice junctions of the transcript are supported by at least one non-suspect mRNA. tsl2 - the best supporting mRNA is flagged as suspect or the support is from multiple ESTs. tsl3 - the only support is from a single EST. tsl4 - the best supporting EST is flagged as suspect. tsl5 - no single transcript supports the model structure. tslNA - the transcript was not analyzed for one of the following reasons: pseudogene annotation, including transcribed pseudogenes.Human leukocyte antigen (HLA) transcript. Immunoglobin gene transcript.  T-cell receptor transcript. Single-exon transcript (will be included in a future version)'),(429,'protein_coverage','Protein Coverage','Protein coverage for this gene derived from geneTree in compara'),(430,'consensus_coverage','Consensus Coverage','Consensus coverage for this gene derived from geneTree in compara'),(431,'has_start_codon','Contains start codon','Translation attribute'),(437,'lncRNACount','lncRNA_Count','Number of lncRNAs'),(438,'ncRNACount','ncRNA_Count','Number of ncRNAs'),(439,'UnclassPTCount','UnclassPT_Count','Number of Unclassified Processed Transcripts'),(444,'noncoding_cnt_m','Misc non coding genes','Number of unclassified (miscellaneous) non coding genes'),(445,'noncoding_rcnt_m','Readthrough misc non coding genes','Number of readthrough unclassified (miscellaneous) non coding genes'),(446,'noncoding_racnt_m','Readthrough misc non coding genes','Number of readthrough unclassified (miscellaneous) non coding genes on alternate sequences'),(447,'noncoding_acnt_l','Long non coding genes','Number of long non coding genes on alternate sequences'),(448,'noncoding_racnt','Readthrough non coding genes','Number of readthrough non coding genes on alternate sequences'),(449,'noncoding_rcnt','Readthrough non coding genes','Number of readthrough non coding genes'),(450,'rseq_mrna_match','RefSeq model genomic seq to mRNA match','This is a transcript attribute that signifies an exact match between the underlying genomic sequence of the RefSeq transcript with the corresponding RefSeq mRNA sequence the model was built from (based on a match between the transcript stable id and an accession in the RefSeq mRNA file). An exact match occurs when the underlying genomic sequence of the model can be perfectly aligned to the mRNA sequence post polyA clipping.'),(451,'rseq_mrna_nonmatch','RefSeq model genomic seq to mRNA non-match','This is a transcript attribute that signifies a non-match between the underlying genomic sequence of the RefSeq transcript with the corresponding RefSeq mRNA sequence the model was built from. A non-match is deemed to have occurred if the underlying genomic sequence does not have a perfect alignment to the mRNA sequence post polyA clipping. It can also signify that no comparison was possible as the model stable id may not have had a corresponding entry in the RefSeq mRNA file (sometimes happens when accessions are retired or changed). When a non-match occurs one or several of the following transcript attributes will also be present to provide more detail on the nature of the non-match: rseq_5p_mismatch, rseq_cds_mismatch, rseq_3p_mismatch, rseq_nctran_mismatch, rseq_no_comparison'),(452,'rseq_5p_mismatch','RefSeq model genomic seq (5\' UTR) to mRNA mismatch','This is a transcript attribute that signifies a mismatch between the underlying genomic sequence of the RefSeq transcript with the corresponding RefSeq mRNA sequence the model was built from. Specifically, there is a mismatch in the 5\' UTR of the RefSeq model. Information about the mismatch can be found in the value field of the transcript attribute.'),(453,'rseq_cds_mismatch','RefSeq model genomic seq (CDS) to mRNA mismatch','This is a transcript attribute that signifies a mismatch between the underlying genomic sequence of the RefSeq transcript with the corresponding RefSeq mRNA sequence the model was built from. Specifically, there is a mismatch in the CDS of the RefSeq model. Information about the mismatch can be found in the value field of the transcript attribute.'),(454,'rseq_3p_mismatch','RefSeq model genomic seq (3\' UTR) to mRNA mismatch','This is a transcript attribute that signifies a mismatch between the underlying genomic sequence of the RefSeq transcript with the corresponding RefSeq mRNA sequence the model was built from. Specifically, there is a mismatch in the 3\' UTR of the RefSeq model. Information about the mismatch can be found in the value field of the transcript attribute.'),(455,'rseq_nctran_mismatch','RefSeq model genomic seq (non-coding) to mRNA mismatch','This is a transcript attribute that signifies a mismatch between the underlying genomic sequence of the RefSeq transcript with the corresponding RefSeq mRNA sequence the model was built from. This is a comparison between the entire underlying genomic sequence of the RefSeq model to the mRNA in the case of RefSeq models that are non-coding. Information about the mismatch can be found in the value field of the transcript attribute.'),(456,'rseq_no_comparison','RefSeq model no comparison made to mRNA','This is a transcript attribute that signifies that no alignment was carried out between the underlying genomic sequence of RefSeq model and a corresponding RefSeq mRNA. The reason for this is generally that no corresponding, unversioned accession was found in the RefSeq mRNA file for the transcript stable id. This sometimes happens when accessions are retired or replaced. A second possibility is that the sequences were too long and problematic to align (though this is rare). The value field gives more information about the reason no comparison was possible.'),(457,'rseq_ens_match_wt','RefSeq model to overlapping Ensembl model whole transcript match','This is a transcript attribute that signifies that for the RefSeq transcript there is an overlapping Ensembl model that is identical across the whole transcript. A whole transcript match is defined as follows: 1) In the case that both models are coding, the transcript, CDS and peptide sequences are all identical and the genomic coordinates of every exon match. 2) In the case that both transcripts are non-coding the transcript sequences and the genomic coordinates of every exon are identical. No comparison is made between a coding and a non-coding transcript. Useful related attributes are: rseq_ens_match_cds and rseq_ens_no_match.'),(458,'rseq_ens_match_cds','RefSeq model to overlapping Ensembl model CDS match','This is a transcript attribute that signifies that for the RefSeq transcript there is an overlapping Ensembl model that is identical across the CDS region only. A CDS match is defined as follows: the CDS and peptide sequences are identical and the genomic coordinates of every translatable exon match. Useful related attributes are: rseq_ens_match_wt and rseq_ens_no_match.'),(459,'rseq_ens_no_match','RefSeq model to overlapping Ensembl model no match','This is a transcript attribute that signifies that for the RefSeq transcript there is no overlapping Ensembl model that is identical across either the whole transcript or the CDS. This is caused by differences between the transcript, CDS or peptide sequences or between the exon genomic coordinates. Useful related attributes are: rseq_ens_match_wt and rseq_ens_match_cds.'),(468,'ccds_transcript','CCDS transcript','This attribute signifies that a transcript has a matching CCDS transcript (the accession of which will be in the value column of the transcript_attrib entry). A match occurs when the genomic coordinates of all coding exons of the transcript are identical to the genomic coordinate of all coding exons in the overlapping CCDS model.'),(469,'SO_accession','SO accession','Sequence Ontology accession'),(470,'SO_term','SO term','Sequence Ontology term'),(471,'display_term','display term','Ensembl display term'),(472,'NCBI_term','NCBI term','NCBI term'),(473,'feature_SO_term','feature SO term','Sequence Ontology term for the associated feature'),(474,'rank','rank','Relative severity of this variation consequence'),(475,'polyphen_prediction','polyphen prediction','PolyPhen-2 prediction'),(476,'sift_prediction','sift prediction','SIFT prediction'),(477,'short_name','Short name','A shorter name for an instance, e.g. a VariationSet'),(478,'dbsnp_clin_sig','dbSNP/ClinVar clinical significance','The clinical significance of a variant as reported by ClinVar and dbSNP'),(479,'dgva_clin_sig','DGVa clinical significance','The clinical significance of a structural variant as reported by DGVa'),(480,'clinvar_clin_sig','ClinVar clinical significance','The clinical significance of a variant as reported by ClinVar'),(481,'prot_func_analysis','Protein function analysis ','The program used to make protein function predictions'),(482,'associated_gene','Associated gene','ID of gene(s) linked by phenotype association'),(483,'risk_allele','Risk allele','Risk allele in phenotype association'),(484,'p_value','P-value','P-value denoting significance of an observed phenotype annotation'),(485,'variation_names','Variation names','Variant ID(s) linked with a phenotype association'),(486,'sample_id','Sample ID','Sample ID for source of phenotype association'),(487,'strain_id','Strain ID','Strain ID for source of phenotype association'),(488,'lod_score','LOD score','Log Of Odds score'),(489,'variance','Variance','Variance statistic'),(490,'inheritance_type','Inheritance type','Inheritance type of a trait'),(491,'external_id','External ID','External identifier for an entity'),(492,'odds_ratio','Odds ratio','Odds ratio used to denote significance of an observed phenotype annotation'),(493,'beta_coef','Beta coefficient','Beta coefficient (or standardized coefficient) used to denote significance of an observed phenotype annotation'),(494,'allele_symbol','Allele symbol','Allele symbol linked with phenotype association'),(495,'allele_accession_id','Allele accession ID','Allele accession ID linked with phenotype association'),(496,'marker_accession_id','Marker accession ID','Marker ID linked with phenotype association'),(497,'evidence','Variant evidence status','Evidence status for a variant'),(498,'review_status','ClinVar review_status','ClinVar review_status for assertation'),(499,'based_on','Evidence type used for protein impact prediction','Evidence type used for a PolyPhen protein impact prediction'),(500,'conservation_score','Sift conservation score','Median conservation va in an alignment used to make a Sift prediction'),(501,'sequence_number','Number of sequences in alignment','Number of protein sequences in the alignment use to make a protein impact prediction'),(502,'otter_truncated','Otter truncated feature','This feature extends beyond the slice, but has been trimmed. (For use in otter client-server communications.)'),(503,'trans_spliced','Trans-spliced transcript','A single RNA transcript derived from multiple precursor mRNAs.'),(505,'genebuild_msu7_tes','TE-related Gene (MSU)','Number of TE-related genes predicted by <a href=\"http://rice.plantbiology.msu.edu\">MSU</a> through a process of automatic and manual curation'),(506,'ibsc_low_confidence','PGSB low-confidence','Number of low-confidence genes annotated by the <A HREF=\"http://pgsb.helmholtz-muenchen.de/plant/barley/index.jsp\">International Barley Sequencing Consortium</A>'),(507,'pubmed_id','PubMed ID','PubMed identifier'),(508,'var_att','variation_attrib','An attribute of a variation'),(509,'ontology_mapping','Ontology Mapping','Method used to link a description to an ontology term'),(510,'enst_refseq_compare','ENST/RefSeq sequence and structural comparisons','Each Ensembl transcript is compared to overlapping RefSeq transcripts. The comparison is only coding-coding or non-coding to non-coding. The calculations are as follows: Foreach Ensembl transcript:  Fetch overlapping RefSeq transcripts  Foreach RefSeq transcript:    1) Check if all exon coordinates match    2) Check if transcript sequences match    3) If both transcripts are protein coding:        a) Check if the CDS exon coordinates match        b) Check if the CDS sequences match        c) Check if the translation sequences match Checks 1 & 2 only are run on pairs of non-coding transcripts, while checks 1, 2, 3a, 3b and 3c are run for pairs of protein coding transcripts. For non-coding models: RefSeq transcript accessions passing checks 1 and 2 will get a line in the value column consisting of all such accessions (separated by \":\") suffixed  by \":whole_transcript\", to indicate these RefSeq accessions have a complete sequence and structural match across the entire transcript. Checks 3a, 3b and 3c are not considered as the transcripts are non-coding For coding models: RefSeq transcript accessions passing all five checks will get a line in the value column consisting of all such accessions (separated by \":\") suffixed  by \":whole_transcript\", to indicate these RefSeq accessions have a complete sequence and structural match across the entire transcript RefSeq transcript accessions passing tests 3a,3b and 3c (but not both test 1 & 2)  will get a line in the value column consisting of all such accessions (separated by \":\") suffixed  by \":cds_only\", to indicate these RefSeq accessions have a sequence and structural match across only the CDS region.'),(511,'rna_gene_biotype','Biotype','Biotype of an RNA gene'),(512,'cmscan_truncated','Truncated','In a cmscan alignment, the end of the gene which is truncated'),(513,'cmscan_accuracy','Accuracy','The accuracy value in a cmscan alignment'),(514,'cmscan_bias','Bias','The bias value in a cmscan alignment'),(515,'cmscan_gc','GC','The GC value in a cmscan alignment'),(516,'cmscan_significant','Significant','The significant value in a cmscan alignment'),(517,'rfam_accession','Accession','Rfam accession');
/*!40000 ALTER TABLE `attrib_type` ENABLE KEYS */;
UNLOCK TABLES;

--
-- Table structure for table `external_db`
--

DROP TABLE IF EXISTS `external_db`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `external_db` (
  `external_db_id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `db_name` varchar(100) NOT NULL,
  `db_release` varchar(255) DEFAULT NULL,
  `status` enum('KNOWNXREF','KNOWN','XREF','PRED','ORTH','PSEUDO') NOT NULL,
  `priority` int(11) NOT NULL,
  `db_display_name` varchar(255) DEFAULT NULL,
  `type` enum('ARRAY','ALT_TRANS','ALT_GENE','MISC','LIT','PRIMARY_DB_SYNONYM','ENSEMBL') DEFAULT NULL,
  `secondary_db_name` varchar(255) DEFAULT NULL,
  `secondary_db_table` varchar(255) DEFAULT NULL,
  `description` text,
  PRIMARY KEY (`external_db_id`),
  UNIQUE KEY `db_name_db_release_idx` (`db_name`,`db_release`)
) ENGINE=MyISAM AUTO_INCREMENT=50836 DEFAULT CHARSET=latin1;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `external_db`
--

LOCK TABLES `external_db` WRITE;
/*!40000 ALTER TABLE `external_db` DISABLE KEYS */;
INSERT INTO `external_db` VALUES (211,'VB_Community_Symbol',NULL,'KNOWN',50,'VB Community Symbol','MISC',NULL,NULL,NULL),(212,'VB_External_Description','1','PRED',50,'VB External Description','MISC',NULL,NULL,NULL),(213,'VB_RNA_Description','1','PRED',50,'VB RNA Description','MISC',NULL,NULL,NULL),(230,'Ribosomal_Protein_Gene_Database','Dec.2010','XREF',50,'Ribosomal Protein Gene DB','MISC',NULL,NULL,NULL),(250,'ImmunoDB','1','KNOWN',50,'ImmunoDB','MISC',NULL,NULL,NULL),(400,'Celera_Gene',NULL,'PRED',5,'Celera gene','MISC',NULL,NULL,NULL),(410,'Celera_Pep',NULL,'PRED',5,'Celera peptide','MISC',NULL,NULL,NULL),(420,'Celera_Trans',NULL,'PRED',5,'Celera transcript','MISC',NULL,NULL,NULL),(700,'EMBL',NULL,'KNOWNXREF',5,'European Nucleotide Archive','MISC',NULL,NULL,NULL),(701,'EMBL_predicted',NULL,'PRED',119,'Predicted European Nucleotide Archive','MISC',NULL,NULL,NULL),(800,'flybase_gene_id',NULL,'KNOWNXREF',5,'FlyBase gene','MISC',NULL,NULL,NULL),(801,'flybase_translation_id',NULL,'KNOWNXREF',5,'FlyBase translation','MISC',NULL,NULL,NULL),(804,'flybase_annotation_id',NULL,'KNOWNXREF',5,'FlyBase annotation','MISC',NULL,NULL,NULL),(805,'flybase_symbol',NULL,'KNOWNXREF',50,'FlyBase symbol','MISC',NULL,NULL,NULL),(808,'flybase_transcript_id',NULL,'KNOWNXREF',299,'FlyBase transcript','MISC',NULL,NULL,NULL),(812,'gadfly_translation_cgid',NULL,'XREF',10,'Gadfly translation CGID','MISC',NULL,NULL,NULL),(821,'FlyBaseName_gene',NULL,'KNOWN',5,'FlyBase gene name','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(824,'FlyBaseName_translation',NULL,'KNOWN',5,'FlyBase translation name','MISC',NULL,NULL,NULL),(825,'FlyBaseName_transcript',NULL,'XREF',50,'FlyBase transcript name','MISC',NULL,NULL,NULL),(826,'FlyBaseCGID_gene',NULL,'XREF',5,'FlyBase gene CGID','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(827,'FlyBaseCGID_transcript',NULL,'XREF',50,'FlyBase transcript CGID','MISC',NULL,NULL,NULL),(828,'FlyBaseCGID_translation',NULL,'XREF',50,'FlyBase translation CGID','MISC',NULL,NULL,NULL),(829,'BioGRID',NULL,'KNOWNXREF',5,'BioGRID Interaction data, The General Repository for Interaction Datasets','MISC',NULL,NULL,NULL),(831,'GenomeRNAi',NULL,'KNOWNXREF',5,'GenomeRNAi, a database for cell-based RNAi phenotypes','MISC',NULL,NULL,NULL),(832,'InteractiveFly',NULL,'KNOWNXREF',5,'The Interactive Fly','MISC',NULL,NULL,NULL),(833,'MitoDrome',NULL,'KNOWNXREF',5,'Nuclear genes encoding proteins targeted to the mitochondrion','MISC',NULL,NULL,NULL),(883,'FlyReactome',NULL,'KNOWNXREF',5,'FlyReactome, a Curated Knowledgebase of Drosophila Melanogaster Pathways','MISC',NULL,NULL,NULL),(1000,'GO','143','XREF',5,'GO','MISC',NULL,NULL,NULL),(1001,'GO_REF',NULL,'KNOWN',5,'GO Reference Collection','LIT',NULL,NULL,NULL),(1100,'HGNC',NULL,'KNOWNXREF',100,'HGNC Symbol','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(1200,'Interpro',NULL,'XREF',5,'InterPro','MISC',NULL,NULL,NULL),(1300,'EntrezGene',NULL,'KNOWNXREF',250,'EntrezGene','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(1400,'MGI',NULL,'KNOWNXREF',270,'MGI Symbol','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(1510,'MIM_GENE',NULL,'KNOWNXREF',5,'MIM gene','MISC',NULL,NULL,NULL),(1520,'MIM_MORBID',NULL,'KNOWNXREF',150,'MIM morbid','MISC',NULL,NULL,NULL),(1600,'PDB',NULL,'KNOWNXREF',5,'PDB','MISC',NULL,NULL,NULL),(1700,'protein_id',NULL,'KNOWNXREF',5,'INSDC protein ID','MISC',NULL,NULL,NULL),(1701,'protein_id_predicted',NULL,'PRED',4,'Predicted INSDC Protein ID','MISC',NULL,NULL,NULL),(1800,'RefSeq_dna',NULL,'KNOWN',200,'RefSeq DNA','MISC',NULL,NULL,NULL),(1801,'RefSeq_mRNA','NCBI Reference Sequence (RefSeq) Database Release 76, May 9, 2016','KNOWN',200,'RefSeq mRNA','MISC',NULL,NULL,NULL),(1803,'RefSeq_ncRNA','NCBI Reference Sequence (RefSeq) Database Release 76, May 9, 2016','KNOWN',200,'RefSeq ncRNA','MISC',NULL,NULL,NULL),(1805,'RefSeq_dna_predicted',NULL,'PRED',199,'RefSeq DNA predicted','MISC',NULL,NULL,NULL),(1806,'RefSeq_mRNA_predicted','NCBI Reference Sequence (RefSeq) Database Release 76, May 9, 2016','PRED',199,'RefSeq mRNA predicted','MISC',NULL,NULL,NULL),(1808,'RefSeq_ncRNA_predicted','NCBI Reference Sequence (RefSeq) Database Release 76, May 9, 2016','PRED',199,'RefSeq ncRNA predicted','MISC',NULL,NULL,NULL),(1810,'RefSeq_peptide','NCBI Reference Sequence (RefSeq) Database Release 76, May 9, 2016','KNOWN',210,'RefSeq peptide','MISC',NULL,NULL,NULL),(1815,'RefSeq_peptide_predicted','NCBI Reference Sequence (RefSeq) Database Release 76, May 9, 2016','PRED',197,'RefSeq peptide predicted','MISC',NULL,NULL,NULL),(1820,'RefSeq_rna',NULL,'KNOWN',230,'RefSeq RNA','MISC',NULL,NULL,NULL),(1825,'RefSeq_rna_predicted',NULL,'PRED',195,'RefSeq RNA predicted','MISC',NULL,NULL,NULL),(1830,'RefSeq_genomic',NULL,'KNOWN',193,'RefSeq Genomic','MISC',NULL,NULL,'This external_db_id can be used in the seq_region_synonym table. For species such as human, cow, dog we store chromosome names (1-22, X,Y) in the name column of the seq_region table. The RefSeq_genomic is stored as a synonym. eg. NC_000011.10, NT_187365.1'),(2000,'Uniprot/SPTREMBL','UniProtKB/TrEMBL Release 2016_05 of 11-May-2016','KNOWN',119,'UniProtKB/TrEMBL','MISC',NULL,NULL,NULL),(2200,'Uniprot/SWISSPROT','UniProtKB/Swiss-Prot Release 2016_05 of 11-May-2016','KNOWN',220,'UniProtKB/Swiss-Prot','MISC',NULL,NULL,NULL),(2202,'Uniprot/Varsplic',NULL,'KNOWN',15,'UniProtKB/SpliceVariant','MISC',NULL,NULL,NULL),(2250,'UniProtKB_all',NULL,'XREF',40,'UniProtKB generic accession number (TrEMBL or SwissProt not differentiated)','MISC',NULL,NULL,NULL),(2300,'Vega_gene',NULL,'KNOWNXREF',5,'Vega gene','MISC',NULL,NULL,NULL),(2310,'Vega_transcript',NULL,'KNOWNXREF',160,'Vega transcript','MISC',NULL,NULL,NULL),(2320,'Vega_translation',NULL,'KNOWNXREF',5,'Vega translation','MISC',NULL,NULL,NULL),(2400,'wormbase_gene',NULL,'KNOWN',5,'WormBase Gene','MISC',NULL,NULL,NULL),(2410,'wormbase_transcript',NULL,'KNOWN',5,'WormBase Transcript','MISC',NULL,NULL,NULL),(2420,'wormpep_id',NULL,'KNOWN',5,'Wormpep ID','MISC',NULL,NULL,NULL),(2440,'wormbase_locus',NULL,'KNOWN',5,'WormBase Locus','MISC',NULL,NULL,NULL),(2510,'ZFIN_ID',NULL,'KNOWNXREF',260,'ZFIN','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(2530,'ZFIN_xpat',NULL,'KNOWNXREF',4,'ZFIN expression patterns','MISC',NULL,NULL,NULL),(2600,'GeneDB',NULL,'KNOWN',5,'GeneDB','MISC',NULL,NULL,NULL),(2601,'cint_jgi_v1',NULL,'KNOWN',10,'JGI Gene ID (JGI 1.0)','MISC',NULL,NULL,NULL),(2602,'cint_jgi_v2',NULL,'KNOWN',5,'JGI Gene ID (JGI 2.0)','MISC',NULL,NULL,NULL),(2610,'cint_aniseed_v1',NULL,'KNOWN',5,'ANISEED (JGI 1.0)','MISC',NULL,NULL,NULL),(2611,'cint_aniseed_v2',NULL,'KNOWN',5,'ANISEED (JGI 2.0)','MISC',NULL,NULL,NULL),(2700,'Ens_Hs_gene',NULL,'XREF',5,'Ensembl Human Gene','MISC',NULL,NULL,NULL),(2710,'Ens_Hs_transcript',NULL,'XREF',170,'Ensembl Human Transcript','MISC',NULL,NULL,NULL),(2720,'Ens_Hs_translation',NULL,'XREF',5,'Ensembl Human Translation','MISC',NULL,NULL,NULL),(2800,'RGD',NULL,'KNOWNXREF',50,'RGD Symbol','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(2900,'Genoscope_pred_gene',NULL,'XREF',38,'Genoscope pred gene','MISC',NULL,NULL,NULL),(2910,'Genoscope_pred_transcript',NULL,'XREF',39,'Genoscope pred transcript','MISC',NULL,NULL,NULL),(2920,'Genoscope_annotated_gene',NULL,'KNOWN',40,'Genoscope annotated gene','MISC',NULL,NULL,NULL),(2930,'Genoscope_pred_translation',NULL,'XREF',40,'Genoscope pred translation','MISC',NULL,NULL,NULL),(2940,'Genoscope_ann_transcript',NULL,'KNOWN',40,'Genoscope annotated transcript','MISC',NULL,NULL,NULL),(2950,'Genoscope_ann_translation',NULL,'KNOWN',40,'Genoscope annotated translation','MISC',NULL,NULL,NULL),(3300,'miRBase',NULL,'KNOWNXREF',180,'miRBase','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(3400,'SGD',NULL,'KNOWN',50,'SGD','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(3500,'IPI',NULL,'KNOWN',5,'IPI','MISC',NULL,NULL,NULL),(3800,'CCDS',NULL,'KNOWN',240,'CCDS','MISC',NULL,NULL,NULL),(3900,'PUBMED',NULL,'KNOWN',5,'Sequence Publications','LIT',NULL,NULL,NULL),(4100,'UniGene','15 Apr 2013 \\t, UniGene Build #195 Rattus norvegicus','KNOWN',5,'UniGene','MISC',NULL,NULL,NULL),(4200,'RFAM',NULL,'XREF',190,'RFAM','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(4400,'Xenbase',NULL,'KNOWN',1,'Xenbase','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(4600,'OTTG',NULL,'XREF',1,'Havana gene','ALT_GENE',NULL,NULL,NULL),(4610,'OTTT',NULL,'XREF',1,'Havana transcript','ALT_TRANS',NULL,NULL,NULL),(4620,'OTTP',NULL,'XREF',1,'Havana translation','MISC',NULL,NULL,NULL),(4670,'shares_CDS_with_OTTT',NULL,'XREF',1,'Havana transcript having same CDS','ALT_TRANS',NULL,NULL,NULL),(4680,'shares_CDS_and_UTR_with_OTTT',NULL,'XREF',1,'Transcript having exact match between ENSEMBL and HAVANA','ALT_TRANS',NULL,NULL,NULL),(4690,'ENSG',NULL,'XREF',1,'Ensembl gene','ALT_GENE',NULL,NULL,NULL),(4700,'Medaka',NULL,'XREF',1,'Medaka Genome Project','MISC',NULL,NULL,NULL),(4810,'ENST_ident',NULL,'XREF',1,'Ensembl transcript having exact match with Havana','ALT_TRANS',NULL,NULL,NULL),(4820,'ENST_CDS',NULL,'XREF',1,'Ensembl transcript sharing CDS with Havana','ALT_TRANS',NULL,NULL,NULL),(5010,'Ens_Mm_transcript',NULL,'XREF',40,'Ensembl Mouse Transcript','MISC',NULL,NULL,NULL),(5020,'Ens_Mm_translation',NULL,'XREF',40,'Ensembl Mouse Translation','MISC',NULL,NULL,NULL),(5100,'IMGT_HLA',NULL,'KNOWN',7,'IMGT/HLA','MISC',NULL,NULL,NULL),(5300,'Oxford_FGU_Md_tscript',NULL,'PRED',1,'Oxford Ponting Group Monodelphis predictions transcript id','MISC',NULL,NULL,'Updated from alt_gene to misc due to user query'),(5400,'Oxford_FGU_Oa_gene',NULL,'PRED',1,'Oxford Ponting Group Platypus predictions gene id','MISC',NULL,NULL,'Updated from alt_gene to misc due to user query'),(5500,'Oxford_FGU_Oa_tscript',NULL,'PRED',1,'Oxford Ponting Group Platypus predictions transcript id','MISC',NULL,NULL,NULL),(5720,'Ens_Cf_translation',NULL,'XREF',5,'Ensembl Dog Translation','MISC',NULL,NULL,NULL),(5810,'Ens_Dr_transcript',NULL,'XREF',5,'Ensembl Zebrafish Transcript','MISC',NULL,NULL,NULL),(5820,'Ens_Dr_translation',NULL,'XREF',5,'Ensembl Zebrafish Translation','MISC',NULL,NULL,NULL),(6400,'Ens_Gg_gene',NULL,'XREF',40,'Ensembl Chicken Gene','MISC',NULL,NULL,NULL),(6410,'Ens_Gg_transcript',NULL,'XREF',40,'Ensembl Chicken Transcript','MISC',NULL,NULL,NULL),(6420,'Ens_Gg_translation',NULL,'XREF',40,'Ensembl Chicken Translation','MISC',NULL,NULL,NULL),(6500,'IMCB_Tr_gene',NULL,'XREF',40,'IMCB Fugu Gene','MISC',NULL,NULL,NULL),(6820,'Ens_Tr_transcript',NULL,'XREF',40,'Ensembl Fugu Transcript','MISC',NULL,NULL,NULL),(6900,'Ens_Ga_translation',NULL,'XREF',5,'Ensembl Stickleback Translation','MISC',NULL,NULL,NULL),(7170,'Ixodes_ManualAnnotation',NULL,'PRED',30,'Ixodes ManualAnnotation','MISC',NULL,NULL,NULL),(7180,'VB_Community_Annotation','1','KNOWN',50,'VB Community Annotation','MISC',NULL,NULL,NULL),(7200,'IMGT/LIGM_DB',NULL,'XREF',10,'IMGT/LIGM-DB','MISC',NULL,NULL,NULL),(7201,'IMGT/GENE_DB',NULL,'KNOWN',40,'IMGT/GENE-DB','MISC',NULL,NULL,NULL),(7300,'Ens_Mg_transcript',NULL,'XREF',40,'Ensembl Turkey Transcript','MISC',NULL,NULL,NULL),(7400,'Ens_Tg_transcript',NULL,'XREF',40,'Ensembl Zebrafinch Transcript','MISC',NULL,NULL,NULL),(7620,'Ens_Fc_translation',NULL,'XREF',40,'Ensembl Cat Translation','MISC',NULL,NULL,NULL),(8100,'Platypus_olfactory_receptor',NULL,'XREF',40,'Platypus olfactory receptor gene','MISC',NULL,NULL,NULL),(8200,'PRF',NULL,'XREF',5,'PRF','MISC',NULL,NULL,NULL),(8300,'PIR',NULL,'XREF',5,'PIR','MISC',NULL,NULL,NULL),(8400,'TetNig_cdna',NULL,'XREF',40,'Genoscope tetraodon cDNA','MISC',NULL,NULL,NULL),(8500,'GasAcu_cdna',NULL,'XREF',40,'Stickleback cDNA','MISC',NULL,NULL,NULL),(8600,'OrnAna_454_cdna',NULL,'XREF',40,'Platypus 454 cDNA','MISC',NULL,NULL,NULL),(8700,'TakRub_cdna',NULL,'XREF',40,'Takifugu cDNA','MISC',NULL,NULL,NULL),(8800,'CioInt_est',NULL,'XREF',40,'Ciona intestinalis EST','MISC',NULL,NULL,NULL),(8900,'CioInt_cdna',NULL,'XREF',40,'Ciona intestinalis cDNA','MISC',NULL,NULL,NULL),(9200,'TakRub_est',NULL,'XREF',40,'Takifugu EST','MISC',NULL,NULL,NULL),(9400,'TetNig_mouse_econtig',NULL,'XREF',40,'Genoscope mouse exofish econtig','MISC',NULL,NULL,NULL),(9410,'TetNig_fugu_econtig',NULL,'XREF',40,'Genoscope fugu exofish econtig','MISC',NULL,NULL,NULL),(9420,'TetNig_human_econtig',NULL,'XREF',40,'Genoscope human exofish econtig','MISC',NULL,NULL,NULL),(9450,'TetNig_rat_econtig',NULL,'XREF',40,'Genoscope rat exofish econtig','MISC',NULL,NULL,NULL),(9460,'TetNig_chick_econtig',NULL,'XREF',40,'Genoscope chicken exofish econtig','MISC',NULL,NULL,NULL),(9500,'OryLat_est',NULL,'XREF',40,'Medaka EST','MISC',NULL,NULL,NULL),(9600,'Trace_archive',NULL,'XREF',40,'Trace Archive id','MISC',NULL,NULL,NULL),(9700,'CioSav_est',NULL,'XREF',40,'Ciona savignyi EST','MISC',NULL,NULL,NULL),(10000,'wormbase_id',NULL,'XREF',30,'Generic exdb for WormBase id of any type for feature table','MISC',NULL,NULL,NULL),(10100,'EPD',NULL,'KNOWNXREF',5,'Eukaryotic Promoter Database (Bucher)','MISC',NULL,NULL,NULL),(10300,'MEROPS',NULL,'KNOWNXREF',5,'MEROPS - the Peptidase Database','MISC',NULL,NULL,NULL),(10500,'TransFac',NULL,'KNOWNXREF',5,'TransFac, database of transcription factors and their binding sites','MISC',NULL,NULL,NULL),(11000,'UCSC',NULL,'KNOWNXREF',100,'UCSC Stable ID','MISC',NULL,NULL,NULL),(12310,'Clone_based_vega_gene',NULL,'KNOWNXREF',5,'Clone-based (Vega)','MISC',NULL,NULL,NULL),(12315,'Clone_based_ensembl_gene',NULL,'XREF',5,'Clone-based (Ensembl)','MISC',NULL,NULL,NULL),(12410,'Clone_based_vega_transcript',NULL,'KNOWNXREF',5,'Clone-based (Vega)','MISC',NULL,NULL,NULL),(12415,'Clone_based_ensembl_transcript',NULL,'XREF',5,'Clone-based (Ensembl)','MISC',NULL,NULL,NULL),(12500,'DBASS3',NULL,'XREF',50,'DataBase of Aberrant 3\' Splice Sites','MISC',NULL,NULL,NULL),(12505,'DBASS5',NULL,'XREF',50,'DataBase of Aberrant 5\' Splice Sites','MISC',NULL,NULL,NULL),(12510,'HPA',NULL,'XREF',50,'Human Protein Atlas','MISC',NULL,NULL,NULL),(12600,'WikiGene',NULL,'XREF',50,'WikiGene','MISC',NULL,NULL,NULL),(12601,'Tgut_symbol',NULL,'KNOWNXREF',100,'Tgut symbol - Prof. David Burt, Roslin Institute and Royal (Dick) School of Veterinary Studies, Edinburgh University, UK','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(12620,'Duck_consortium',NULL,'XREF',0,'Duck consortium','MISC',NULL,NULL,NULL),(12630,'BGI_duck_transcriptome',NULL,'XREF',0,'Beijing Genomics Institute (BGI) duck transcriptome','MISC',NULL,NULL,NULL),(12700,'goslim_goa',NULL,'XREF',5,'GOSlim GOA','MISC',NULL,NULL,NULL),(12710,'WTSI_gorilla_transcriptome',NULL,'XREF',0,'Wellcome Trust Sanger Institute (WTSI) gorilla transcriptome','MISC',NULL,NULL,NULL),(20005,'UniParc',NULL,'KNOWNXREF',0,'UniParc','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20008,'BRENDA',NULL,'KNOWNXREF',0,'BRENDA','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20009,'BioCyc',NULL,'KNOWNXREF',0,'BioCyc','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20013,'EchoBASE',NULL,'KNOWNXREF',0,'EchoBASE','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20014,'EcoGene',NULL,'KNOWNXREF',0,'EcoGene','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20040,'SubtiList',NULL,'KNOWNXREF',0,'SubtiList','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20043,'TubercuList',NULL,'KNOWNXREF',0,'TubercuList','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20046,'ArrayExpress',NULL,'XREF',0,'Expression Atlas','MISC',NULL,NULL,NULL),(20059,'DIP',NULL,'KNOWNXREF',0,'DIP','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20061,'DisProt',NULL,'KNOWNXREF',0,'DisProt','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20062,'DrugBank',NULL,'KNOWNXREF',0,'DrugBank','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20066,'HAMAP',NULL,'KNOWNXREF',0,'HAMAP','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20071,'HSSP',NULL,'KNOWNXREF',0,'HSSP','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20072,'IntAct',NULL,'KNOWNXREF',0,'IntAct','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20074,'KEGG',NULL,'KNOWNXREF',0,'KEGG','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20078,'PDBsum',NULL,'KNOWNXREF',0,'PDBsum','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20082,'PeroxiBase',NULL,'KNOWNXREF',0,'PeroxiBase','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20085,'REBASE',NULL,'KNOWNXREF',0,'REBASE','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20088,'Reactome','54','KNOWNXREF',0,'Reactome','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20097,'RegulonDb_Transcript',NULL,'KNOWNXREF',0,'RegulonDb','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20101,'IntEnz',NULL,'KNOWNXREF',0,'EnzymeCommission','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20200,'EMBLBANK_GENE',NULL,'KNOWNXREF',117,'European Nucleotide Archive','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20215,'phatr_jgi_v2',NULL,'KNOWNXREF',0,'JGI ID (JGI 2.0)','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20216,'phatr_jgi_v2_bd_gene',NULL,'KNOWNXREF',0,'JGI Gene ID (JGI 2.0 unmapped sequence)','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20217,'phatr_jgi_v2_gene',NULL,'KNOWNXREF',0,'JGI Gene ID (JGI 2.0)','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20220,'thaps_jgi_v2_bd',NULL,'KNOWNXREF',0,'JGI ID (JGI 2.0 unmapped sequence)','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20221,'thaps_jgi_v2',NULL,'KNOWNXREF',0,'JGI ID (JGI 2.0)','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20228,'CADRE_Afum_A1163',NULL,'KNOWNXREF',0,'CADRE','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20301,'RNAMMER',NULL,'KNOWNXREF',0,'RNAmmer','MISC',NULL,NULL,NULL),(20302,'TRNASCAN_SE',NULL,'KNOWNXREF',0,'tRNAScan-SE','MISC',NULL,NULL,NULL),(20303,'dictyBase',NULL,'KNOWNXREF',0,'DictyBase','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20306,'ASPGD',NULL,'KNOWNXREF',0,'AspGD','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20307,'ASPGD_GENE',NULL,'KNOWNXREF',0,'AspGD','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20308,'ASPGD_TRANSCRIPT',NULL,'KNOWNXREF',0,'AspGD','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20309,'CADRE',NULL,'KNOWNXREF',0,'CADRE','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20310,'CADRE_GENE',NULL,'KNOWNXREF',0,'CADRE','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20311,'CADRE_TRANSCRIPT',NULL,'KNOWNXREF',0,'CADRE','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20312,'SGD_GENE',NULL,'KNOWNXREF',0,'SGD','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20313,'SGD_TRANSCRIPT',NULL,'KNOWNXREF',0,'SGD','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20317,'phyra_jgi_v1.1','1.1','KNOWNXREF',0,'JGI ID (JGI 1.1)','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20320,'physo1_jgi_v1.1','1.1','KNOWNXREF',0,'JGI ID (JGI 1.1)','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20325,'BROAD_P_infestans','1','KNOWNXREF',0,'Broad Institute','MISC',NULL,NULL,NULL),(20329,'triad_jgi_v1.0_gene','1','KNOWNXREF',0,'JGI Gene ID (JGI Triad 1.0)','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20330,'triad_jgi_v1.0_transcript','1','KNOWNXREF',0,'JGI Transcript ID (JGI Triad 1.0)','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20331,'triad_jgi_v1.0','1','KNOWNXREF',0,'JGI ID (JGI Triad 1.0)','MISC',NULL,NULL,NULL),(20334,'SCHISTODB','1','KNOWNXREF',0,'SchistoDB','MISC',NULL,NULL,NULL),(20337,'BROAD_F_oxysporum','1','KNOWNXREF',0,'Broad Institute','MISC',NULL,NULL,NULL),(20338,'BROAD_F_oxysporum_GENE','1','KNOWNXREF',0,'Broad Institute','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20339,'BROAD_F_oxysporum_TRANSCRIPT','1','KNOWNXREF',0,'Broad Institute','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20340,'BROAD_G_zeae','1','KNOWNXREF',0,'Broad Institute','MISC',NULL,NULL,NULL),(20343,'BROAD_G_moniliformis','1','KNOWNXREF',0,'Broad Institute','MISC',NULL,NULL,NULL),(20344,'BROAD_G_moniliformis_GENE','1','KNOWNXREF',0,'Broad Institute','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20345,'BROAD_G_moniliformis_TRANSCRIPT','1','KNOWNXREF',0,'Broad Institute','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20348,'PGD_GENE','1','KNOWNXREF',0,'Pythium genome database','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20354,'BROAD_U_maydis','1','KNOWNXREF',0,'Broad Institute','MISC',NULL,NULL,NULL),(20355,'BROAD_U_maydis_GENE','1','KNOWNXREF',0,'Broad Institute','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20356,'BROAD_U_maydis_TRANSCRIPT','1','KNOWNXREF',0,'Broad Institute','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20357,'BROAD_P_triticina','1','KNOWNXREF',0,'Broad Institute','MISC',NULL,NULL,NULL),(20358,'BROAD_P_triticina_GENE','1','KNOWNXREF',0,'Broad Institute','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20359,'BROAD_P_triticina_TRANSCRIPT','1','KNOWNXREF',0,'Broad Institute','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20360,'SpBase','1','KNOWNXREF',0,'SpBase','MISC',NULL,NULL,NULL),(20362,'necha_jgi_v2.0_gene','2','KNOWNXREF',0,'Nectria JGI gene','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20363,'necha_jgi_v2.0_transcript','2','KNOWNXREF',0,'Nectria JGI transcript','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20364,'necha_jgi_v2.0','2','KNOWNXREF',0,'Nectria JGI','MISC',NULL,NULL,NULL),(20365,'tuber_genoscope_gene','1','KNOWNXREF',0,'Tuber Genome Browser','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20368,'Mycgr3_jgi_v2.0_gene','2','KNOWNXREF',0,'Mycosphaerella JGI gene','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20372,'BROAD_P_NODORUM_GENE','1','KNOWNXREF',0,'Broad Institute','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20377,'PomBase_TRANSCRIPT',NULL,'KNOWNXREF',0,'PomBase','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20378,'PomBase_TRANSLATION','1','KNOWNXREF',0,'PomBase','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20380,'CBS',NULL,'KNOWNXREF',0,'CBS Server','MISC',NULL,NULL,NULL),(20381,'COG',NULL,'KNOWNXREF',0,'Cluster of Orthologous Groups (NCBI)','MISC',NULL,NULL,NULL),(20382,'CGD',NULL,'KNOWN',0,'Candida Genome Database','MISC',NULL,NULL,NULL),(20383,'ProSite',NULL,'XREF',0,'ProSite ID','MISC',NULL,NULL,NULL),(20385,'SMART',NULL,'XREF',0,'SMART ID','MISC',NULL,NULL,NULL),(20386,'SPD',NULL,'KNOWNXREF',0,'SPD ID','MISC',NULL,NULL,NULL),(20388,'PBO',NULL,'KNOWNXREF',0,'PomBase Ontology ID','MISC',NULL,NULL,NULL),(20389,'FYPO',NULL,'KNOWNXREF',0,'Fission Yeast Phenotype Ontology','MISC',NULL,NULL,NULL),(20390,'MOD',NULL,'KNOWNXREF',0,'PSI-MOD, Post Translational Modification Ontology','MISC',NULL,NULL,NULL),(20391,'AGD_GENE','1','KNOWNXREF',0,'Ashbya Genome Database','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(20392,'AGD_TRANSCRIPT','1','KNOWNXREF',0,'Ashbya Genome Database','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(50000,'BGI_EST_Cluster',NULL,'XREF',0,'EST Cluster (BGI Indica)','MISC',NULL,NULL,NULL),(50005,'EC_NUMBER',NULL,'XREF',0,'Enzyme EC Number','MISC',NULL,NULL,NULL),(50010,'FSTs',NULL,'XREF',0,'Flanking Sequence Tag (FST)','MISC',NULL,NULL,NULL),(50020,'Gramene_Pathway',NULL,'XREF',0,'Gramene Pathway','MISC',NULL,NULL,NULL),(50035,'Gramene_MarkersDB',NULL,'XREF',0,'Gramene','MISC',NULL,NULL,NULL),(50040,'Gramene_MarkersDB_mRNA',NULL,'XREF',0,'GenBank mRNA (Gramene)','MISC',NULL,NULL,NULL),(50050,'Gramene_MarkersDB_EST',NULL,'XREF',0,'GenBank EST (Gramene)','MISC',NULL,NULL,NULL),(50090,'NASC_GENE_ID',NULL,'XREF',1,'NASC Gene ID','MISC',NULL,NULL,NULL),(50110,'PlantGDB_PUT',NULL,'XREF',1,'PlantGDB Transcript','MISC',NULL,NULL,NULL),(50120,'RAP_Gene',NULL,'XREF',0,'IRGSP/RAP Gene','MISC',NULL,NULL,NULL),(50140,'TIGR_LOCUS',NULL,'XREF',1,'MSU/TIGR Locus','MISC',NULL,NULL,NULL),(50160,'TAIR_LOCUS',NULL,'XREF',8,'TAIR','MISC',NULL,NULL,NULL),(50170,'TAIR_LOCUS_MODEL',NULL,'XREF',8,'TAIR Locus (Model)','MISC',NULL,NULL,NULL),(50180,'TIGR_GeneIndex',NULL,'XREF',0,'Gene Index','MISC',NULL,NULL,NULL),(50185,'TO',NULL,'XREF',0,'Plant Trait Ontology','MISC',NULL,NULL,NULL),(50190,'PO',NULL,'XREF',0,'Plant Structure Ontology','MISC',NULL,NULL,NULL),(50196,'GO_to_gene',NULL,'XREF',0,'Gene Ontology','MISC',NULL,NULL,NULL),(50200,'GRO',NULL,'XREF',0,'Plant Growth Stage','MISC',NULL,NULL,NULL),(50541,'LRG',NULL,'KNOWN',10,'Locus Reference Genomic','MISC',NULL,NULL,NULL),(50542,'ENS_LRG_gene',NULL,'KNOWN',10,'LRG display in Ensembl','MISC',NULL,NULL,NULL),(50543,'ENS_LRG_transcript',NULL,'KNOWN',10,'LRG display in Ensembl','MISC',NULL,NULL,NULL),(50607,'miRBase_trans_name',NULL,'KNOWNXREF',5,'miRBase transcript name','MISC',NULL,NULL,'transcript name from miRBase'),(50608,'RFAM_trans_name',NULL,'KNOWNXREF',5,'RFAM transcript name','MISC',NULL,NULL,'transcript name from RFAM'),(50609,'HGNC_trans_name',NULL,'KNOWNXREF',5,'HGNC transcript name','MISC',NULL,NULL,'transcript name from HGNC'),(50610,'MGI_trans_name',NULL,'KNOWNXREF',5,'MGI transcript name','MISC',NULL,NULL,'transcript name from MGI'),(50611,'ZFIN_ID_trans_name',NULL,'KNOWNXREF',5,'ZFIN transcript name','MISC',NULL,NULL,'transcript name from ZFIN'),(50620,'Illumina_Bodymap_transcriptome',NULL,'XREF',0,'Illumina Bodymap transcriptome','MISC',NULL,NULL,NULL),(50621,'Uniprot_gn',NULL,'KNOWN',116,'UniProtKB Gene Name','MISC',NULL,NULL,NULL),(50622,'GGSC',NULL,'KNOWNXREF',0,'Gibbon Genome Sequencing Consortium','MISC',NULL,NULL,'Gibbon Genome Sequencing Consortium'),(50623,'Broad Institute',NULL,'KNOWN',5,'Broad Institute of MIT and Harvard','MISC',NULL,NULL,NULL),(50624,'PFAM',NULL,'XREF',0,'PFAM','MISC',NULL,NULL,'Pfam from Vega'),(50625,'SGSC',NULL,'KNOWN',5,'Swine Genome Sequencing Consortium','MISC',NULL,NULL,NULL),(50626,'RIKEN',NULL,'KNOWN',5,'Riken Institute','MISC',NULL,NULL,NULL),(50627,'SHSC',NULL,'KNOWNXREF',0,'Tasmanian Devil Sequencing Consortium','MISC',NULL,NULL,'Tasmanian Devil Genome Sequencing Consortium'),(50628,'PFAM','22','XREF',0,'PFAM','MISC',NULL,NULL,'Pfam release 22, from Vega'),(50629,'PFAM','23','XREF',0,'PFAM','MISC',NULL,NULL,'Pfam release 23, from Vega'),(50630,'PFAM','24','XREF',0,'PFAM','MISC',NULL,NULL,'Pfam release 24, from Vega'),(50631,'PFAM','25','XREF',0,'PFAM','MISC',NULL,NULL,'Pfam release 25, from Vega'),(50632,'PFAM','26','XREF',0,'PFAM','MISC',NULL,NULL,'Pfam release 26, from Vega'),(50633,'GRC_primary_assembly',NULL,'XREF',0,'GRC primary assembly','MISC',NULL,NULL,'Primary assembly of this species genome from the GRC'),(50634,'Ens_Ga_gene',NULL,'XREF',5,'Ensembl Stickleback Gene','MISC',NULL,NULL,NULL),(50635,'Ens_Ga_transcript',NULL,'XREF',170,'Ensembl Stickleback Transcript','MISC',NULL,NULL,NULL),(50637,'CSAC',NULL,'KNOWN',5,'Chimpanzee Sequencing and Analysis Consortium','MISC',NULL,NULL,NULL),(50640,'PomBase_Interaction_GENETIC','1','KNOWNXREF',5,'PomBase Interaction Genetic','MISC',NULL,NULL,NULL),(50641,'PomBase_Interaction_PHYSICAL','1','KNOWNXREF',5,'PomBase Interaction Physical','MISC',NULL,NULL,NULL),(50642,'PomBase_Gene_Name','1','KNOWNXREF',5,'PomBase Gene Name','MISC',NULL,NULL,NULL),(50643,'PomBase_Systematic_ID','1','KNOWNXREF',5,'PomBase Systematic ID','MISC',NULL,NULL,NULL),(50644,'PomBase_Ortholog','1','KNOWNXREF',5,'Orthologous Gene','MISC',NULL,NULL,NULL),(50645,'Ens_Ss_translation',NULL,'XREF',5,'Ensembl Pig Translation','MISC',NULL,NULL,NULL),(50646,'Psinensis_sequencing_consortium',NULL,'KNOWN',5,'Pelodiscus sinensis sequencing consortium','MISC',NULL,NULL,NULL),(50647,'BROAD_Magnaporthe_DB','1','KNOWNXREF',0,'Magnaporthe comparative DB','MISC',NULL,NULL,'Broad Institute Magnaporthe comparative Database'),(50648,'BROAD_Magnaporthe_DB_GENE','1','KNOWNXREF',0,'Magnaporthe comparative DB','PRIMARY_DB_SYNONYM',NULL,NULL,'Broad Institute Magnaporthe comparative Database'),(50649,'BROAD_Magnaporthe_DB_TRANSCRIPT','1','KNOWNXREF',0,'Magnaporthe comparative DB','PRIMARY_DB_SYNONYM',NULL,NULL,'Broad Institute Magnaporthe comparative Database'),(50650,'BGI Shenzhen',NULL,'KNOWN',5,'Beijing Genomics Institute, Shenzhen','MISC',NULL,NULL,NULL),(50651,'ICGSC',NULL,'KNOWN',5,'International Cat Genome Sequencing Consortium','MISC',NULL,NULL,NULL),(50653,'WashingtonUniversity',NULL,'KNOWN',5,'The Genome Institute, Washington University School of Medicine','MISC',NULL,NULL,NULL),(50654,'BDGP_cDNA','1','KNOWN',50,'Fly cDNA','MISC',NULL,NULL,'cDNA from the Berkeley Drosophila Genome Project'),(50655,'BDGP_cDNA_Gold_Collection','1','KNOWN',50,'Fly gold cDNA','MISC',NULL,NULL,'Gold Collection cDNA from the Berkeley Drosophila Genome Project'),(50656,'Ens_Tg_translation',NULL,'KNOWN',5,'Ensembl Zebrafinch Translation','MISC',NULL,NULL,NULL),(50658,'ToxoDB','1','KNOWNXREF',1,'ToxoDB','PRIMARY_DB_SYNONYM',NULL,NULL,'Toxoplasma Genomic Resource Database.'),(50660,'SO','1','KNOWNXREF',0,'Sequence Ontology','MISC',NULL,NULL,'Sequence Ontology'),(50661,'BRAD_GENE','1','XREF',5,'BrassEnsembl','MISC',NULL,NULL,'The Brassica Genome Database'),(50662,'BRAD_TRANSCRIPT','1','XREF',5,'BrassEnsembl','MISC',NULL,NULL,'The Brassica Genome Database'),(50663,'Quick_Go',NULL,'KNOWN',1,'GOA','MISC',NULL,NULL,'Resource to show GOA records for Uniprot Accessions'),(50664,'Quick_Go_Evidence',NULL,'KNOWN',1,'Quick Go','MISC',NULL,NULL,'Resource to show GOA view of original annotation sources'),(50666,'IGDBCAS-v1.4_GENE','1.4','KNOWN',1,'IGDBCAS v1.4 Gene','MISC',NULL,NULL,NULL),(50667,'IGDBCAS-v1.4_TRANSCRIPT','1.4','KNOWN',1,'IGDBCAS v1.4 Transcript','MISC',NULL,NULL,NULL),(50668,'JGI_GENE','1','XREF',1,'JGI Gene','MISC',NULL,NULL,'gene ID/NAME in the JGI gff3 file'),(50669,'JGI_TRANSCRIPT','1','XREF',1,'JGI Transcript','MISC',NULL,NULL,'transcript/mRNA ID/NAME in the JGI gff3 file'),(50672,'IRGSPv1_GENE','1','KNOWNXREF',1,'IRGSP v1 Gene','MISC',NULL,NULL,NULL),(50673,'IRGSPv1_TRANSCRIPT','1','KNOWNXREF',1,'IRGSP v1 Transcript','MISC',NULL,NULL,NULL),(50674,'Ens_Lc_gene',NULL,'XREF',5,'Ensembl Coelacanth Gene','MISC',NULL,NULL,NULL),(50677,'RGSC',NULL,'KNOWNXREF',0,'Rat Genome Sequencing Consortium','MISC',NULL,NULL,'Rat Genome Sequencing Consortium'),(50678,'BROAD_coelacanth_transcriptome',NULL,'XREF',0,'BROAD Institute coelacanth transcriptome','MISC',NULL,NULL,NULL),(50679,'Ens_Ac_translation',NULL,'XREF',40,'Ensembl Anole Lizard Translation','MISC',NULL,NULL,NULL),(50682,'BGI_Gene','1','XREF',50,'BGI Gene','MISC',NULL,NULL,'BGI gene identifier'),(50683,'GeneIndex','1','XREF',50,'GeneIndex','MISC',NULL,NULL,NULL),(50684,'EO','1','XREF',0,'Environment Ontology','MISC',NULL,NULL,'Plant environmental conditions ontology terms.\r\nMore information in there:\r\nhttp://www.gramene.org/plant_ontology/ontology_browse.html#eo'),(50687,'Uppsala University',NULL,'KNOWN',5,'Uppsala University','MISC',NULL,NULL,NULL),(50688,'GenBank','1','XREF',1,'GenBank','MISC',NULL,NULL,'GenBank identifier'),(50690,'RefSeq_gene_name',NULL,'KNOWN',200,'RefSeq gene name','MISC',NULL,NULL,'This is read out of the *.gff files provided to us by the NCBI when we do a RefSeq import or a CCDS comparison. It has the key \"gene_id\" and is found at the start of the 9th field.'),(50691,'SGN','1','KNOWNXREF',1,'Sol Genomics Network','MISC',NULL,NULL,NULL),(50692,'GRC_alignment_import',NULL,'XREF',0,'GRC alignment import','MISC',NULL,NULL,'Imported alignments from the Genome Reference Consortium.'),(50693,'BROAD_monodelphis_transcriptome',NULL,'XREF',0,'Broad Institute transcriptome data for monodelphis','MISC',NULL,NULL,NULL),(50694,'SIB_mondelphis_transcriptome',NULL,'XREF',0,'Swiss Institute of Bioinformatics transcriptome data for monodelphis','MISC',NULL,NULL,NULL),(50695,'SIB_platypus_transcriptome',NULL,'XREF',0,'Swiss Institute of Bioinformatics transcriptome data for platypus','MISC',NULL,NULL,NULL),(50696,'SIB_orangutan_transcriptome',NULL,'XREF',0,'Swiss Institute of Bioinformatics transcriptome data for orangutan','MISC',NULL,NULL,NULL),(50697,'NCBI_TAXONOMY','1','KNOWNXREF',1,'NCBI species','MISC',NULL,NULL,'Organism listed in NCBI'),(50698,'PHI','1','KNOWNXREF',1,'PHI-base accession','MISC',NULL,NULL,'Parasite-host interaction identifiers from PHI-base'),(50700,'AnoXcel','2','KNOWN',50,'AnoXcel','MISC',NULL,NULL,'VectorBase xref.'),(50701,'PGSC_GENE','1','XREF',1,'PGSC','MISC',NULL,NULL,'Potato Genome Sequencing Consortium database'),(50703,'Uniprot_gn_trans_name',NULL,'KNOWNXREF',5,'UniProt transcript name','MISC',NULL,NULL,'transcript name projected from UniProt gene name'),(50708,'GMGC_GENE','1','KNOWNXREF',1,'GMGC','PRIMARY_DB_SYNONYM',NULL,NULL,'Global Musa Genomics Consortium gene cross-references'),(50709,'tamu_vibs',NULL,'KNOWN',0,'TAMU VIBS','MISC',NULL,NULL,'The Texas A&M College of Veterinary Medicine & Biomedical Sciences, Department of Veterinary Integrative Biosciences.'),(50710,'INSDC','0','KNOWNXREF',5,'International Nucleotide Sequence Database Collaboration','MISC','INSDC',NULL,'This external_db_id was initially made for use in the seq_region_synonym table. For species such as human, cow, dog we store chromosome names (1-22, X,Y) in the name column of the seq_region table. The INSDC accession is stored as a synonym.'),(50711,'ISGC',NULL,'KNOWNXREF',0,'International Sheep Genome Consortium','MISC',NULL,NULL,'International Sheep Genome Consortium'),(50713,'TAIR_SYMBOL','10','KNOWN',60,'TAIR Gene Name','PRIMARY_DB_SYNONYM',NULL,NULL,'TAIR gene names'),(50715,'TIGR_ID','1','XREF',1,'TIGR Internal ID','MISC',NULL,NULL,'cross-references to TIGR internal identifiers, attached to genes.'),(50717,'GOA','1','KNOWNXREF',1,'GOA','MISC',NULL,NULL,'GOA xrefs (to UniProt identifiers)'),(50718,'ENA_GENE','1','KNOWNXREF',1,'European Nucleotide Archive','MISC',NULL,NULL,'Cross-references to the source gene model when importing from ENA.'),(50719,'ENA_TRANSCRIPT','1','KNOWNXREF',1,'European Nucleotide Archive','MISC',NULL,NULL,'Cross-references to the source transcript model when importing from ENA.'),(50720,'LGBPratt_EST_Cluster','1','KNOWNXREF',1,'LGBPratt_EST_Cluster','MISC',NULL,NULL,NULL),(50721,'BaylorCollege',NULL,'KNOWN',5,'Baylor College of Medicine, Texas Medical Center','MISC',NULL,NULL,NULL),(50722,'Rhea','1','KNOWNXREF',1,'Rhea','MISC',NULL,NULL,'Cross-references to Rhea attached to translation objects'),(50723,'Ens_Bt_translation',NULL,'XREF',5,'Ensembl Cow Translation','MISC',NULL,NULL,NULL),(50726,'ENA_FEATURE_GENE','1','KNOWNXREF',1,'ENA','MISC',NULL,NULL,'Cross-reference to the ENA object from which the gene has been inferred, when loading a genome from ENA records.'),(50727,'ENA_FEATURE_TRANSCRIPT','1','KNOWNXREF',1,'ENA','MISC',NULL,NULL,'Cross-reference to the ENA object from which the transcript has been inferred, when loading a genome from ENA records.'),(50728,'ENA_FEATURE_PROTEIN','1','KNOWNXREF',1,'ENA','MISC',NULL,NULL,'Cross-reference to the ENA object from which the translation has been inferred, when loading a genome from ENA records.'),(50730,'AGP',NULL,'KNOWN',5,'Anolis Genome Project','MISC',NULL,NULL,NULL),(50731,'Turkey Genome Consortium',NULL,'KNOWN',5,'Turkey Genome Consortium','MISC',NULL,NULL,'Turkey Genome Consortium'),(50732,'Yutaka_Satou_Kyoto_University',NULL,'KNOWN',5,'Yutaka Satou Kyoto University','MISC',NULL,NULL,'Yutaka Satou Kyoto University'),(50733,'Chicken_Genome_Consortium',NULL,'KNOWN',5,'International Chicken Genome Consortium','MISC',NULL,NULL,NULL),(50734,'TAIR_TRANSLATION','1','XREF',1,'TAIR Translation identifier','MISC',NULL,NULL,'TAIR identifier to link to Ensembl Translation entities.\r\nThe main requirement behind this entry, is to be able to link TAIR GO annotations to Ensembl Translation objects.'),(50735,'AGI_GENE','1','XREF',1,'AGI Gene','MISC',NULL,NULL,'annotation provided by Arizona Genome Institute'),(50736,'AGI_TRANSCRIPT','1','XREF',1,'AGI Transcript','MISC',NULL,NULL,'annotation provided by Arizona Genome Institute'),(50737,'CGNC',NULL,'KNOWNXREF',100,'CGNC Symbol','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(50738,'MetaCyc','1','KNOWNXREF',1,'MetaCyc','MISC',NULL,NULL,'MetaCyc database of metabolic pathways'),(50739,'PHIP','1','KNOWNXREF',1,'PHI-base phenotype ontology','MISC',NULL,NULL,'PHI-base phenotype ontology for genes proven to affect the outcome of pathogen-host interactions'),(50740,'PHIE','1','KNOWNXREF',1,'PHI-base condition ontology','MISC',NULL,NULL,'PHI-base experimental evidence ontology for genes proven to affect the outcome of pathogen-host interactions'),(50742,'PubMedCentral',NULL,'KNOWN',5,'PubMedCentral','LIT',NULL,NULL,'PMC identifier for use with PubMedCentral (http://www.ncbi.nlm.nih.gov/pmc and http://europepmc.org/)'),(50743,'PFAM','27','XREF',0,'PFAM','MISC',NULL,NULL,'Pfam release 27, from Vega'),(50744,'EntrezGene_trans_name',NULL,'KNOWNXREF',250,'EntrezGene transcript name','MISC',NULL,NULL,'transcript name projected from EntrezGene gene name'),(50745,'BROAD_M_violaceum','1','KNOWNXREF',0,'Broad Institute','MISC',NULL,NULL,'Broad Institute Microbotryum violaceum Database'),(50746,'BROAD_M_violaceum_GENE','1','KNOWNXREF',0,'Broad Institute','PRIMARY_DB_SYNONYM',NULL,NULL,'Broad Institute Microbotryum violaceum Database'),(50747,'BROAD_M_violaceum_TRANSCRIPT','1','KNOWNXREF',0,'Broad Institute','PRIMARY_DB_SYNONYM',NULL,NULL,'Broad Institute Microbotryum violaceum Database'),(50748,'Ensembl_Plants','1','XREF',1,'Ensembl Plants','ENSEMBL',NULL,NULL,'Ensembl Plants identifiers serving GO projections from a reference species to other plants species.'),(50749,'ChEMBL',NULL,'KNOWNXREF',5,'ChEMBL','MISC',NULL,NULL,'ChEMBL external references imported via Uniprot mappings'),(50750,'IBSC','1','KNOWNXREF',50,'IBSC Gene ID','MISC',NULL,NULL,'International Barley Sequencing Consortium gene identifier'),(50752,'PRIDE',NULL,'XREF',0,'PRIDE','MISC',NULL,NULL,NULL),(50753,'LegioList',NULL,'XREF',0,'LegioList','MISC',NULL,NULL,'Database dedicated to the analysis of the genomes of Legionella'),(50755,'BindingDB',NULL,'XREF',0,'BindingDB','MISC',NULL,NULL,'BindingDB is a public, web-accessible database of measured binding affinities'),(50756,'STRING',NULL,'XREF',0,'STRING','MISC',NULL,NULL,'STRING - Known and Predicted Protein-Protein Interactions'),(50757,'UniPathway',NULL,'XREF',0,'UniPathway','MISC',NULL,NULL,'UniPathway is a manually curated resource of enzyme-catalyzed and spontaneous chemical reactions.'),(50759,'SWISS_MODEL',NULL,'XREF',0,'SWISS-MODEL','MISC',NULL,NULL,'The SWISS-MODEL Repository is a database of annotated three-dimensional comparative protein structure models generated by the fully automated homology-modelling pipeline SWISS-MODEL.'),(50762,'CYGD',NULL,'XREF',0,'CYGD','MISC',NULL,NULL,'The MIPS Comprehensive Yeast Genome Database (CYGD) aims to present information on the molecular structure and functional network of the entirely sequenced, well-studied model eukaryote, the budding yeast Saccharomyces cerevisiae.'),(50770,'PseudoCAP',NULL,'XREF',0,'Pseudomonas Genome Database','MISC',NULL,NULL,'The Pseudomonas Genome database has been a resource for peer-reviewed, continually-updated annotations for the Pseudomonas aeruginosa PAO1 reference strain\'s genome and, more recently, comparative analyses of several related Pseudomonas species.'),(50771,'MINT',NULL,'XREF',0,'MINT','MISC',NULL,NULL,'MINT, the Molecular INTeraction database, focuses on experimentally verified protein-protein interactions mined from the scientific literature by expert curators.'),(50772,'Ant Genomes Portal',NULL,'KNOWNXREF',0,'Ant Genomes Portal','MISC',NULL,NULL,NULL),(50775,'Ens_Mg_translation',NULL,'XREF',40,'Ensembl Turkey Translation','MISC',NULL,NULL,NULL),(50776,'LNCC',NULL,'XREF',30,'LNCC','MISC',NULL,NULL,'The National Laboratory of Scientific Computation (Laboratorio Nacional de Computacao Cientifica: LNCC), Brazil'),(50777,'RGD_trans_name',NULL,'KNOWNXREF',50,'RGD transcript name','MISC',NULL,NULL,'transcript name from RGD'),(50778,'Xenbase_trans_name',NULL,'KNOWNXREF',1,'Xenbase transcript name','MISC',NULL,NULL,'transcript name from Xenbase'),(50779,'NasoniaBase',NULL,'KNOWN',0,'NasoniaBase','MISC',NULL,NULL,NULL),(50780,'wormbase_gseqname',NULL,'KNOWN',6,'WormBase Gene Sequence-name','MISC',NULL,NULL,NULL),(50782,'Sanger',NULL,'KNOWN',5,'Sanger Wellcome Trust Institute','MISC',NULL,NULL,NULL),(50783,'ENSP_ident',NULL,'XREF',1,'Ensembl translation having exact match with Havana','ALT_TRANS',NULL,NULL,NULL),(50784,'Ensembl_Metazoa','1','XREF',1,'Ensembl Metazoa','ENSEMBL',NULL,NULL,'Ensembl Metazoa identifiers serving GO projections from a reference species to other metazoa species.'),(50785,'Ens_Mmu_translation',NULL,'XREF',40,'Ensembl Macaque Translation','MISC',NULL,NULL,NULL),(50786,'PomBase_GO_AnnotationExtensions',NULL,'KNOWN',1,'PomBase GO AnnotationExtensions','MISC',NULL,NULL,'Specific accession used by PomBase for ontology annoation extensions'),(50788,'SP_SL',NULL,'KNOWN',0,'UniProtKB/SwissProt Subcellular Locatisation','MISC',NULL,NULL,'UniProt Subcellular Location (http://www.uniprot.org/manual/subcellular_location)'),(50789,'PR',NULL,'KNOWNXREF',0,'Protein Ontology','MISC',NULL,NULL,'Protein Ontology term (previously PR) - http://pir.georgetown.edu/pro/pro.shtml'),(50790,'PUBMED_POMBASE',NULL,'KNOWN',1,'PomBase PubMed ID','LIT',NULL,NULL,'PomBase specific publications'),(50792,'CL',NULL,'KNOWNXREF',0,'Cell Type Ontology','MISC',NULL,NULL,'Cell type ontology annotations (http://cellontology.org/)'),(50793,'ChEBI',NULL,'KNOWNXREF',0,'ChEBI Chemical Database','MISC',NULL,NULL,'ChEBI chemical database'),(50794,'FYPO_EXT',NULL,'KNOWNXREF',1,'FYPO Extensions','MISC',NULL,NULL,'Annotation extension terms for the Fission Yeast Phenotype Ontology (FYPO)'),(50795,'EcoCyc',NULL,'KNOWNXREF',1,'EcoCyc','MISC',NULL,NULL,'Scientific database for E. coli and a division of MetaCyc (http://ecocyc.org/)'),(50796,'PATO',NULL,'KNOWNXREF',1,'Phenotypic Qualities Ontology','MISC',NULL,NULL,'http://obofoundry.org/wiki/index.php/PATO:Main_Page'),(50797,'UniProtKB-KW',NULL,'KNOWNXREF',0,'UniProtKB Keywords','MISC',NULL,NULL,'UniProtKB Keywords for describing UniProt annotations'),(50798,'PBQ',NULL,'KNOWNXREF',0,'PomBase Quiescence Ontology','MISC',NULL,NULL,'Used in the annotation of quiescent cell phenotypes.'),(50799,'University of Texas',NULL,'KNOWN',5,'University of Texas','MISC',NULL,NULL,NULL),(50801,'KEGG_Enzyme',NULL,'XREF',0,'KEGG','MISC',NULL,NULL,'KEGG pathway and enzyme'),(50802,'gfbgp',NULL,'KNOWN',0,'Glanville Fritillary Butterfly Genome Project','MISC',NULL,NULL,NULL),(50803,'ensembl_internal_synonym',NULL,'KNOWNXREF',100,'Ensembl internal synonym','MISC',NULL,NULL,NULL),(50808,'DOI',NULL,'KNOWN',5,'Digital Object Identifier','LIT',NULL,NULL,'Digital Object Identifier used to uniquely identify an object such as an electronic document.'),(50809,'zngp',NULL,'KNOWNXREF',0,'Zootermopsis nevadensis Genome Project','MISC',NULL,NULL,NULL),(50814,'FGDB',NULL,'KNOWNXREF',0,'FGDB','ALT_GENE',NULL,NULL,'MIPS Fusarium graminearum Genome Database'),(50815,'Ensembl_Fungi','1','XREF',1,'Ensembl Fungi','ENSEMBL',NULL,NULL,'Ensembl Fungi identifiers serving GO projections from a reference species to other fungal species.'),(50817,'RNACentral',NULL,'KNOWNXREF',0,'RNACentral','PRIMARY_DB_SYNONYM',NULL,NULL,'Sequence identity to RNACentral ids'),(50818,'Smar_Community_Annotation',NULL,'KNOWN',50,'Strigamia maritima Community Annotation','MISC',NULL,NULL,NULL),(50819,'RFAM_GENE',NULL,'KNOWN',0,'Rfam Gene Name','MISC',NULL,NULL,'Name for gene assigned from Rfam alignment'),(50820,'PGSB_WHEAT_GENE','2.2','KNOWNXREF',1,'Wheat survey sequence PGSB gene identifier','MISC',NULL,NULL,'Bread wheat survey sequence gene models from PGSB Wheat Genome Database (formerly MIPS). Release 2.2'),(50821,'GDEC_GENE','1.0','KNOWNXREF',1,'GDEC 3B gene identifier','MISC',NULL,NULL,'Bread wheat 3B pseudomolecule gene models from INRA GDEC group.\n'),(50822,'ARAPORT_GENE','1','KNOWNXREF',0,'Araport','MISC',NULL,NULL,'Gene identifiers from Araport, the Arabidopsis information portal (formerly TAIR).'),(50823,'MTGD','4.0','KNOWNXREF',0,'MTGD','PRIMARY_DB_SYNONYM',NULL,NULL,'The Medicago truncatula Genome Database (MTGD), maintained by JCVI.'),(50824,'Reactome_gene',NULL,'KNOWNXREF',0,'Reactome','PRIMARY_DB_SYNONYM',NULL,NULL,'Reactome entries linked to genes'),(50825,'Reactome_transcript',NULL,'KNOWNXREF',0,'Reactome','PRIMARY_DB_SYNONYM',NULL,NULL,'Reactome entries linked to transcripts'),(50826,'PB_REF',NULL,'KNOWNXREF',1,'PomBase Reference','MISC',NULL,NULL,'Specific accession used by PomBase for source reference'),(50827,'GEO',NULL,'XREF',0,'Gene Expression Omnibus','MISC',NULL,NULL,NULL),(50828,'GigaDB_A_tauschii','1','KNOWNXREF',1,'Aegilops tauschii GigaDB identifier','MISC',NULL,NULL,'Aegilops tauschii GigaDB gene identifier'),(20350,'RegulonDb_Operon',NULL,'KNOWNXREF',0,'RegulonDb','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(50830,'Gramene_Plant_Reactome',NULL,'XREF',0,'Gramene Plant Reactome','MISC',NULL,NULL,'Plant Reactome is a database of plant metabolic and regulatory pathways'),(50831,'EFO',NULL,'KNOWNXREF',0,'Experimental Factor Ontology','MISC',NULL,NULL,'The Experimental Factor Ontology provides a systematic description of many experimental variables available in EBI databases, and for external projects such as the NHGRI GWAS catalog. It combines parts of several biological ontologies, such as UBERON anatomy, ChEBI chemical compounds, and Cell Ontology.'),(50832,'EpiRR',NULL,'KNOWN',0,'EpiRR','MISC',NULL,NULL,'EpiRR is a registry of reference epigenomes.'),(50833,'VGNC',NULL,'KNOWNXREF',100,'VGNC Symbol','PRIMARY_DB_SYNONYM',NULL,NULL,NULL),(50834,'VGNC_trans_name',NULL,'KNOWNXREF',5,'VGNC transcript name','MISC',NULL,NULL,'transcript name from VGNC'),(50741,'dbSNP',NULL,'XREF',101,'dbSNP','MISC',NULL,NULL,'dbSNP variant corresponding to Vega 1kG LOF Variant.'),(50835,'VectorBase_Proteome',NULL,'KNOWN',50,'VectorBase Proteome','MISC',NULL,NULL,NULL);
/*!40000 ALTER TABLE `external_db` ENABLE KEYS */;
UNLOCK TABLES;

--
-- Table structure for table `misc_set`
--

DROP TABLE IF EXISTS `misc_set`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `misc_set` (
  `misc_set_id` smallint(5) unsigned NOT NULL AUTO_INCREMENT,
  `code` varchar(25) NOT NULL DEFAULT '',
  `name` varchar(255) NOT NULL DEFAULT '',
  `description` text NOT NULL,
  `max_length` int(10) unsigned NOT NULL,
  PRIMARY KEY (`misc_set_id`),
  UNIQUE KEY `code_idx` (`code`)
) ENGINE=MyISAM AUTO_INCREMENT=91 DEFAULT CHARSET=latin1;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `misc_set`
--

LOCK TABLES `misc_set` WRITE;
/*!40000 ALTER TABLE `misc_set` DISABLE KEYS */;
INSERT INTO `misc_set` VALUES (1,'encode','Encode regions','Regions that were examined in the <a rel=\"external\" href=\"http://www.genome.gov/10005107\">Encode</a> pilot project.',1877426),(2,'genomicctg','Genomic contigs','The largest contigs constructed by the Genome Reference Consortium, <a rel=\"external\" href=\"http://genomereference.org\">GRC</a>, for this genomic assembly.',100530253),(3,'tilepath','Tilepath','BAC clones upon which the current genomic assembly was based.  Colours indicate a finished sequence (gold clones), phase 1 sequencing (dark pink), phase 0 (light pink) or other (grey).  Clones that have been fluorescence in situ hybridisation (FISH) mapped are marked with a black triangle in the top left corner.  If a clone is shown in outline, the mapping to the genome is problematic and the true length may not be displayed.',1071649),(9,'encode_excluded','ENCODE excluded regions','Genomic regions that have been identified as problematic in the human genome by the ENCODE project. These regions contain abnormally high unique-mapping read coverage signal in open chromatin and control/input datasets across many cell lines and are classified as dispersed high mappability artifacts, dispersed low mappability artifacts, satellite repeats, centromeric repeats, telomeric repeats or snRNA repeats.',4132352),(10,'bac_map','BAC map','Full list of mapped BAC clones',500000),(11,'ASSEMBLY','WGS/Clones assembly','The tiling path of clones and WGS contigs',1000000),(12,'bacs','BACs','BACs',13738670),(13,'gap','gaps','',44494),(14,'matepairs','Mate pairs','Inconsistent mate pairs',21557510),(15,'atag_CLE','Assembly tag','Clone_left_end',1000000),(16,'atag_CRE','Assembly tag','Clone_right_end',1000000),(17,'atag_Misc','Assembly tag','Misc',1000000),(18,'atag_Unsure','Assembly tag','Unsure',1000000),(19,'CHORI-17','CHORI-17 library','Mapped clone end sequences from human CHORI-17 BAC library (CH17)',18090412),(20,'RPCI-11','RPCI-11 library','Mapped clone end sequences from human RPCI-11 BAC library (RP11)',18084827),(21,'ABC','ABC libraries','Mapped clone end sequences from ABCs (ABC7, ABC8, ABC9, ABC10, ABC11, ABC12, ABC13, ABC14, ABC16, ABC21, ABC22, ABC24, ABC27) human fosmid libraries (ABC)',18047227),(22,'WIBR-2','WIBR-2 library','Mapped clone end sequences from WIBR-2 human fosmid library (WI2)',18043599),(23,'CH243','CHORI-243 library','Mapped clone end sequences from sheep CHORI-243 BAC library (CH243)',197393133),(59,'RP71_clones','RP71 zebrafish clones','RP71 library of clones for zebrafish',0),(60,'DKEYP_clones','DKEYP zebrafish clones','DKEYP library of clones for zebrafish',0),(61,'DKEY_clones','DKEY zebrafish clones','DKEY library of clones for zebrafish',0),(62,'CH211_clones','CH211 zebrafish clones','CH211 library of clones for zebrafish',0),(63,'CH73_clones','CH73 zebrafish clones','CH73 library of clones for zebrafish',0),(64,'ZFOS_clones','ZFOS zebrafish clones','ZFOS library of clones for zebrafish',0),(65,'CH1073_clones','CH1073 zebrafish clones','CH1073 library of clones for zebrafish',0),(70,'CH26_clones','CH26 mouse clones','CH26 library of clones for mouse',0),(71,'CH28_clones','CH28 mouse clones','CH28 library of clones for mouse',0),(72,'CH33_clones','CH33 mouse clones','CH33 library of clones for mouse',0),(73,'CH36_clones','CH36 mouse clones','CH36 library of clones for mouse',0),(74,'CH34_clones','CH34 mouse clones','CH34 library of clones for mouse',0),(75,'CT7_clones','CT7 mouse clones','CT7 library of clones for mouse',0),(76,'CH25_clones','CH25 mouse clones','CH25 library of clones for mouse',0),(77,'MM_DBa_clones','MM_DBa mouse clones','MM_DBa library of clones for mouse',0),(78,'RP22_clones','RP22 mouse clones','RP22 library of clones for mouse',0),(79,'RP21_clones','RP21 mouse clones','RP21 library of clones for mouse',0),(80,'MHPP_clones','MHPP mouse clones','MHPP library of clones for mouse',0),(81,'C3H_clones','C3H mouse clones','C3H library of clones for mouse',0),(82,'RP24_clones','RP24 mouse clones','RP24 library of clones for mouse',0),(83,'bMQ_clones','bMQ mouse clones','bMQ library of clones for mouse',0),(84,'CH29_clones','CH29 mouse clones','CH29 library of clones for mouse',0),(85,'MHPN_clones','MHPN mouse clones','MHPN library of clones for mouse',0),(86,'RP23_clones','RP23 mouse clones','RP23 library of clones for mouse',0),(87,'MSMg01_clones','MSMg01 mouse clones','MSMg01 library of clones for mouse',0),(88,'B6Ng01_clones','B6Ng01 mouse clones','B6Ng01 library of clones for mouse',0),(89,'DN_clones','DN mouse clones','DN library of clones for mouse',0),(90,'WI1_clones','WI1 mouse clones','WI1 library of clones for mouse',0);
/*!40000 ALTER TABLE `misc_set` ENABLE KEYS */;
UNLOCK TABLES;

--
-- Table structure for table `unmapped_reason`
--

DROP TABLE IF EXISTS `unmapped_reason`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `unmapped_reason` (
  `unmapped_reason_id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `summary_description` varchar(255) DEFAULT NULL,
  `full_description` varchar(255) DEFAULT NULL,
  PRIMARY KEY (`unmapped_reason_id`)
) ENGINE=MyISAM AUTO_INCREMENT=139 DEFAULT CHARSET=latin1;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `unmapped_reason`
--

LOCK TABLES `unmapped_reason` WRITE;
/*!40000 ALTER TABLE `unmapped_reason` DISABLE KEYS */;
INSERT INTO `unmapped_reason` VALUES (1,'Marker matches multiple times','Marker aligns to the genome > 3 times'),(2,'Marker does not align','Unable to align to the genome'),(3,'Failed to find Stable ID','Stable ID that this xref was linked to no longer exists'),(4,'No mapping done','No mapping done for this type of xref'),(5,'Failed to match','Unable to match to any ensembl entity at all'),(6,'Failed to match at thresholds','Unable to match at the thresholds of 90% for the query or 90% for the target'),(7,'No Master','The dependent xref was not matched due to there being no master xref'),(8,'Master failed','The dependent xref was not matched due to the master xref not being mapped'),(40,'Was not best match','Did not top best transcript match score (0.75)'),(41,'Was not best match','Did not top best transcript match score (0.76)'),(42,'Was not best match','Did not top best transcript match score (0.77)'),(43,'Was not best match','Did not top best transcript match score (0.78)'),(44,'Was not best match','Did not top best transcript match score (0.79)'),(45,'Was not best match','Did not top best transcript match score (0.80)'),(46,'Was not best match','Did not top best transcript match score (0.81)'),(47,'Was not best match','Did not top best transcript match score (0.82)'),(48,'Was not best match','Did not top best transcript match score (0.83)'),(49,'Was not best match','Did not top best transcript match score (0.84)'),(50,'Was not best match','Did not top best transcript match score (0.85)'),(51,'Was not best match','Did not top best transcript match score (0.86)'),(52,'Was not best match','Did not top best transcript match score (0.87)'),(53,'Was not best match','Did not top best transcript match score (0.88)'),(54,'Was not best match','Did not top best transcript match score (0.89)'),(55,'Was not best match','Did not top best transcript match score (0.90)'),(56,'Was not best match','Did not top best transcript match score (0.91)'),(57,'Was not best match','Did not top best transcript match score (0.92)'),(58,'Was not best match','Did not top best transcript match score (0.93)'),(59,'Was not best match','Did not top best transcript match score (0.94)'),(60,'Was not best match','Did not top best transcript match score (0.95)'),(61,'Was not best match','Did not top best transcript match score (0.96)'),(62,'Was not best match','Did not top best transcript match score (0.97)'),(63,'Was not best match','Did not top best transcript match score (0.98)'),(64,'Was not best match','Did not top best transcript match score (0.99)'),(65,'Was not best match','Did not top best transcript match score (1.00)'),(66,'Did not meet threshold','Match score for transcript lower than threshold (0.75)'),(67,'No overlap','No coordinate overlap with any Ensembl transcript'),(96,'Failed to match at thresholds','Unable to match at the thresholds of 100% for the query or 100% for the target'),(125,'Failed to match at thresholds','Unable to match at the thresholds of 55% for the query or 55% for the target'),(126,'>10% N-strings','More than 10% of the sequence consists of strings of Ns. Sequences are not rejected for this reason but this may explain a low coverage hit'),(127,'All long introns','Every intron in these hits is of length 250000-400000bp, we require at least one intron to be shorter than 250000bp'),(128,'GSS sequence','This cDNA has been excluded from the analysis because it is in the GSS (Genome Survey Sequence) division of GenBank'),(129,'Low coverage','Coverage of the best alignment is less than 90% - see query_score for coverage'),(130,'Low coverage with long intron','Hits containing introns longer than 250000bp are rejected if coverage is less than 98% - see query_score for coverage'),(131,'Low percent_id with long intron','Hits containing introns longer than 250000bp are rejected if percentage identity is less than 98% - see query_score for percent_id'),(132,'Low percent_id','Percentage identity of the best alignment is less than 97% - see query_score for percent_id'),(133,'No output from Exonerate','Exonerate returned no hits using standard parameters plus options --maxintron 400000 and --softmasktarget FALSE'),(134,'Parent xref failed to match','Unable to match as parent xref was not mapped'),(135,'Processed pseudogene','Rejected as a processed pseudogene because there are multiple-exon hits with the same coverage which have been rejected for other reasons'),(136,'See kill-list database','This sequence has been excluded from the analysis - see the kill-list database for further details'),(137,'Failed to match at thresholds','Unable to match at the thresholds of 99% for the query or 99% for the target'),(138,'Marker matches multiple times','Marker aligns to the genome > 5 times');
/*!40000 ALTER TABLE `unmapped_reason` ENABLE KEYS */;
UNLOCK TABLES;
/*!40103 SET TIME_ZONE=@OLD_TIME_ZONE */;

/*!40101 SET SQL_MODE=@OLD_SQL_MODE */;
/*!40014 SET FOREIGN_KEY_CHECKS=@OLD_FOREIGN_KEY_CHECKS */;
/*!40014 SET UNIQUE_CHECKS=@OLD_UNIQUE_CHECKS */;
/*!40101 SET CHARACTER_SET_CLIENT=@OLD_CHARACTER_SET_CLIENT */;
/*!40101 SET CHARACTER_SET_RESULTS=@OLD_CHARACTER_SET_RESULTS */;
/*!40101 SET COLLATION_CONNECTION=@OLD_COLLATION_CONNECTION */;
/*!40111 SET SQL_NOTES=@OLD_SQL_NOTES */;

-- Dump completed on 2016-06-26  8:05:18
